// @salt-ds/core/dist-es/menu/MenuGroup.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useId } from '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './MenuGroup.css.js';

const withBaseName = makePrefixer("saltMenuGroup");
const MenuGroup = forwardRef(
  function MenuGroup2(props, ref) {
    const { className, children, label, ...rest } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-menu-group",
      css: css_248z,
      window: targetWindow
    });
    const labelId = useId();
    return /* @__PURE__ */ jsxs(
      "div",
      {
        "aria-labelledby": label ? labelId : void 0,
        className: clsx(withBaseName(), className),
        role: "group",
        ref,
        ...rest,
        children: [
          label && /* @__PURE__ */ jsx("div", { "aria-hidden": true, className: withBaseName("label"), id: labelId, children: label }),
          children
        ]
      }
    );
  }
);

export { MenuGroup };
