// @salt-ds/core/dist-es/menu/MenuPanel.js
import { jsx } from 'react/jsx-runtime';
import { FloatingList } from '@floating-ui/react';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useFloatingComponent } from '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { useMenuContext } from './MenuContext.js';
import { MenuPanelBase } from './MenuPanelBase.js';
import { MenuPanelContext } from './MenuPanelContext.js';

const withBaseName = makePrefixer("saltMenuPanel");
const MenuPanel = forwardRef(
  function MenuPanel2(props, ref) {
    const { children, className, ...rest } = props;
    const { Component: FloatingComponent } = useFloatingComponent();
    const {
      getItemProps,
      openState,
      getFloatingProps,
      refs,
      getPanelPosition,
      context,
      elementsRef,
      activeIndex,
      setFocusInside,
      isNested
    } = useMenuContext();
    const handleRef = useForkRef(ref, refs == null ? void 0 : refs.setFloating);
    return /* @__PURE__ */ jsx(
      MenuPanelContext.Provider,
      {
        value: { activeIndex, getItemProps, setFocusInside },
        children: /* @__PURE__ */ jsx(FloatingList, { elementsRef, children: /* @__PURE__ */ jsx(
          FloatingComponent,
          {
            open: openState,
            role: "menu",
            ...getFloatingProps(),
            ...getPanelPosition(),
            className: clsx(withBaseName(), className),
            focusManagerProps: context ? {
              context,
              initialFocus: isNested ? -1 : 0,
              returnFocus: !isNested,
              modal: false
            } : void 0,
            tabIndex: -1,
            ref: handleRef,
            ...rest,
            children: /* @__PURE__ */ jsx(MenuPanelBase, { children })
          }
        ) })
      }
    );
  }
);

export { MenuPanel };
