// @salt-ds/core/dist-es/tooltip/useTooltip.js
import { offset, shift, flip, arrow, limitShift, useInteractions, useHover, useFocus, useRole, useDismiss, safePolygon } from '@floating-ui/react';
import { useRef } from 'react';
import 'react/jsx-runtime';
import 'clsx';
import { useControlled } from '../utils/useControlled.js';
import { useFloatingUI } from '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { useAriaAnnounce } from './useAriaAnnounce.js';

function useTooltip(props) {
  const {
    enterDelay,
    disabled,
    leaveDelay,
    open: openProp,
    onOpenChange,
    placement: placementProp,
    disableHoverListener,
    disableFocusListener
  } = props ?? {};
  const arrowRef = useRef(null);
  const [open, setOpen] = useControlled({
    controlled: openProp,
    default: false,
    name: "Tooltip",
    state: "open"
  });
  const handleOpenChange = (open2) => {
    setOpen(open2);
    onOpenChange == null ? void 0 : onOpenChange(open2);
  };
  const { floating, reference, x, y, strategy, placement, context, elements } = useFloatingUI({
    open: disabled ? false : open,
    onOpenChange: handleOpenChange,
    placement: placementProp,
    middleware: [
      offset(8),
      shift({ limiter: limitShift() }),
      flip({
        fallbackAxisSideDirection: "end",
        fallbackStrategy: "initialPlacement"
      }),
      arrow({ element: arrowRef })
    ]
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useHover(context, {
      delay: {
        open: enterDelay,
        close: leaveDelay
      },
      enabled: !(disableHoverListener || disabled),
      handleClose: safePolygon()
    }),
    useFocus(context, { enabled: !(disableFocusListener || disabled) }),
    useRole(context, { role: "tooltip" }),
    useDismiss(context, { enabled: !disabled }),
    useAriaAnnounce(context, {
      delay: {
        open: enterDelay,
        close: leaveDelay
      }
    })
  ]);
  const arrowProps = {
    ref: arrowRef,
    context
  };
  const getTooltipProps = () => {
    const { tabIndex, ...tooltipProps } = getFloatingProps({
      // @ts-expect-error - `data-*` props need extra typing when not used on a DOM element.
      "data-placement": placement,
      ref: floating
    });
    return tooltipProps;
  };
  const getTriggerProps = () => getReferenceProps({
    ref: reference
  });
  const getTooltipPosition = () => {
    var _a, _b;
    return {
      top: y ?? 0,
      left: x ?? 0,
      position: strategy,
      width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
      height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight
    };
  };
  return {
    arrowProps,
    open,
    floating,
    reference,
    getTooltipProps,
    getTriggerProps,
    getTooltipPosition
  };
}

export { useTooltip };
