// @salt-ds/core/dist-es/text/Text.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './Text.css.js';

const withBaseName = makePrefixer("saltText");
const Text = forwardRef(function Text2({
  as,
  children,
  className,
  disabled = false,
  maxRows,
  style,
  styleAs,
  variant,
  color: colorProp,
  ...restProps
}, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-text",
    css: css_248z,
    window: targetWindow
  });
  const Component = as ?? "div";
  const textStyles = { "--text-max-rows": maxRows, ...style };
  const color = variant ?? colorProp ?? "primary";
  return /* @__PURE__ */ jsx(
    Component,
    {
      className: clsx(
        withBaseName(),
        {
          [withBaseName("disabled")]: disabled,
          [withBaseName("lineClamp")]: maxRows,
          [withBaseName(styleAs)]: styleAs,
          [withBaseName(color)]: color !== "inherit"
        },
        className
      ),
      ...restProps,
      ref,
      style: textStyles,
      children
    }
  );
});

export { Text };
