// @salt-ds/core/dist-cjs/slider/internal/SliderTooltip.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var Text = require('../../text/Text.js');
require('../../text/Code.js');
require('../../text/Display.js');
require('../../text/Headings.js');
require('../../text/Label.js');
require('../../text/TextAction.js');
require('../../text/TextNotation.js');
require('react');
var makePrefixer = require('../../utils/makePrefixer.js');
require('../../utils/useFloatingUI/useFloatingUI.js');
require('../../utils/useId.js');
require('../../salt-provider/SaltProvider.js');
require('../../viewport/ViewportProvider.js');
var SliderTooltip$1 = require('./SliderTooltip.css.js');

const withBaseName = makePrefixer.makePrefixer("saltSliderTooltip");
const SliderTooltip = ({ value, open }) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-slider-tooltip",
    css: SliderTooltip$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      "aria-hidden": true,
      className: clsx(withBaseName(), {
        [withBaseName("visible")]: open
      }),
      "data-testid": "sliderTooltip",
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(
          "svg",
          {
            className: withBaseName("arrow"),
            "aria-hidden": "true",
            viewBox: "0 1 14 14",
            children: [
              /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M0,0 H14 L7,7 Q7,7 7,7 Z",
                  stroke: " var(--salt-container-primary-background)"
                }
              ),
              /* @__PURE__ */ jsxRuntime.jsx(
                "path",
                {
                  d: "M0,0 H14 L7,7 Q7,7 7,7 Z",
                  stroke: "var(--salt-container-primary-borderColor)"
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(Text.Text, { className: withBaseName("text"), children: value })
      ]
    }
  );
};

exports.SliderTooltip = SliderTooltip;
