// @salt-ds/core/dist-cjs/grid-item/GridItem.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var BreakpointProvider = require('../breakpoints/BreakpointProvider.js');
var parseSpacing = require('../flex-layout/parseSpacing.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
var useResponsiveProp = require('../utils/useResponsiveProp.js');
var GridItem$1 = require('./GridItem.css.js');

const GRID_ALIGNMENT_BASE = [
  "start",
  "end",
  "center",
  "stretch"
];
const withBaseName = makePrefixer.makePrefixer("saltGridItem");
const colStart = "auto";
const colEnd = "auto";
const rowStart = "auto";
const rowEnd = "auto";
const GridItem = React.forwardRef(
  ({
    as,
    children,
    className,
    margin = 0,
    padding = 0,
    colSpan = "auto",
    rowSpan = "auto",
    horizontalAlignment = "stretch",
    verticalAlignment = "stretch",
    style,
    ...rest
  }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-grid-item",
      css: GridItem$1,
      window: targetWindow
    });
    const { matchedBreakpoints } = BreakpointProvider.useBreakpoint();
    const Component = as || "div";
    const gridItemColSpan = useResponsiveProp.resolveResponsiveValue(colSpan, matchedBreakpoints);
    const gridItemRowSpan = useResponsiveProp.resolveResponsiveValue(rowSpan, matchedBreakpoints);
    const gridItemMargin = useResponsiveProp.resolveResponsiveValue(margin, matchedBreakpoints);
    const gridItemPadding = useResponsiveProp.resolveResponsiveValue(padding, matchedBreakpoints);
    const gridColumnStart = gridItemColSpan ? `span ${gridItemColSpan}` : colStart;
    const gridColumnEnd = gridItemColSpan ? `span ${gridItemColSpan}` : colEnd;
    const gridRowStart = gridItemRowSpan ? `span ${gridItemRowSpan}` : rowStart;
    const gridRowEnd = gridItemRowSpan ? `span ${gridItemRowSpan}` : rowEnd;
    const gridStyles = {
      "--gridItem-margin": parseSpacing.parseSpacing(gridItemMargin),
      "--gridItem-padding": parseSpacing.parseSpacing(gridItemPadding),
      ...style,
      "--gridItem-justifySelf": horizontalAlignment,
      "--gridItem-alignSelf": verticalAlignment,
      "--gridItem-gridRowStart": gridRowStart,
      "--gridItem-gridColumnStart": gridColumnStart,
      "--gridItem-gridRowEnd": gridRowEnd,
      "--gridItem-gridColumnEnd": gridColumnEnd
    };
    return /* @__PURE__ */ jsxRuntime.jsx(
      Component,
      {
        className: clsx.clsx(withBaseName(), className),
        style: gridStyles,
        ref,
        ...rest,
        children
      }
    );
  }
);

exports.GRID_ALIGNMENT_BASE = GRID_ALIGNMENT_BASE;
exports.GridItem = GridItem;
