// @salt-ds/core/dist-cjs/file-drop-zone/FileDropZone.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var utils = require('./internal/utils.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var FileDropZone$1 = require('./FileDropZone.css.js');

const withBaseName = makePrefixer.makePrefixer("saltFileDropZone");
const FileDropZone = React.forwardRef(
  function FileDropZone2({
    status,
    className,
    children,
    disabled,
    onDragOver,
    onDragLeave,
    onDrop,
    ...restProps
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-file-drop-zone",
      css: FileDropZone$1,
      window: targetWindow
    });
    const [isActive, setActive] = React.useState(false);
    const regionRef = React.useRef(null);
    const handleRef = useForkRef.useForkRef(ref, regionRef);
    const handleDragOver = (event) => {
      event.preventDefault();
      event.stopPropagation();
      if (disabled) {
        if (event.dataTransfer) {
          event.dataTransfer.dropEffect = "none";
        }
        return;
      }
      if (event.dataTransfer) {
        event.dataTransfer.dropEffect = "copy";
      }
      if (!isActive && utils.containsFiles(event)) {
        setActive(true);
      }
      onDragOver == null ? void 0 : onDragOver(event);
    };
    const handleDragLeave = (event) => {
      if (disabled) {
        return;
      }
      const region = regionRef == null ? void 0 : regionRef.current;
      const eventTarget = event.relatedTarget;
      if (eventTarget !== region && !(region == null ? void 0 : region.contains(eventTarget))) {
        setActive(false);
      }
      onDragLeave == null ? void 0 : onDragLeave(event);
    };
    const handleDrop = (event) => {
      if (disabled) {
        return;
      }
      event.preventDefault();
      const files = utils.extractFiles(event);
      setActive(false);
      onDrop == null ? void 0 : onDrop(event, files);
    };
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: clsx.clsx(
          withBaseName(),
          {
            [withBaseName(status)]: status,
            [withBaseName("active")]: isActive,
            [withBaseName("disabled")]: disabled
          },
          className
        ),
        "aria-disabled": disabled,
        onDragLeave: handleDragLeave,
        onDragOver: handleDragOver,
        onDrop: handleDrop,
        ref: handleRef,
        ...restProps,
        children
      }
    );
  }
);

exports.FileDropZone = FileDropZone;
