// @salt-ds/core/dist-cjs/tag/Tag.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
var clsx = require('clsx');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Tag$1 = require('./Tag.css.js');

const withBaseName = makePrefixer.makePrefixer("saltTag");
const Tag = React.forwardRef(function Tag2({ children, className, variant = "primary", category = 1, bordered, ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-tag",
    css: Tag$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      ref,
      className: clsx.clsx(
        withBaseName(),
        withBaseName(`category-${category}`),
        withBaseName(variant),
        {
          [withBaseName("bordered")]: bordered
        },
        className
      ),
      ...rest,
      children
    }
  );
});

exports.Tag = Tag;
