// @salt-ds/core/dist-cjs/menu/MenuGroup.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useId = require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var MenuGroup$1 = require('./MenuGroup.css.js');

const withBaseName = makePrefixer.makePrefixer("saltMenuGroup");
const MenuGroup = React.forwardRef(
  function MenuGroup2(props, ref) {
    const { className, children, label, ...rest } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-menu-group",
      css: MenuGroup$1,
      window: targetWindow
    });
    const labelId = useId.useId();
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "div",
      {
        "aria-labelledby": label ? labelId : void 0,
        className: clsx.clsx(withBaseName(), className),
        role: "group",
        ref,
        ...rest,
        children: [
          label && /* @__PURE__ */ jsxRuntime.jsx("div", { "aria-hidden": true, className: withBaseName("label"), id: labelId, children: label }),
          children
        ]
      }
    );
  }
);

exports.MenuGroup = MenuGroup;
