// @salt-ds/core/dist-cjs/scrim/Scrim.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Scrim$1 = require('./Scrim.css.js');

const withBaseName = makePrefixer.makePrefixer("saltScrim");
const Scrim = React.forwardRef(function Scrim2({ className, children, fixed = false, open = true, ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-scrim",
    css: Scrim$1,
    window: targetWindow
  });
  if (!open) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: clsx.clsx(
        withBaseName(),
        {
          [withBaseName("fixed")]: fixed
        },
        className
      ),
      "data-testid": "scrim",
      ref,
      ...rest,
      children
    }
  );
});

exports.Scrim = Scrim;
