// @salt-ds/core/dist-es/grid-item/GridItem.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { useBreakpoint } from '../breakpoints/BreakpointProvider.js';
import { parseSpacing } from '../flex-layout/parseSpacing.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import { resolveResponsiveValue } from '../utils/useResponsiveProp.js';
import css_248z from './GridItem.css.js';

const GRID_ALIGNMENT_BASE = [
  "start",
  "end",
  "center",
  "stretch"
];
const withBaseName = makePrefixer("saltGridItem");
const colStart = "auto";
const colEnd = "auto";
const rowStart = "auto";
const rowEnd = "auto";
const GridItem = forwardRef(
  ({
    as,
    children,
    className,
    margin = 0,
    padding = 0,
    colSpan = "auto",
    rowSpan = "auto",
    horizontalAlignment = "stretch",
    verticalAlignment = "stretch",
    style,
    ...rest
  }, ref) => {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-grid-item",
      css: css_248z,
      window: targetWindow
    });
    const { matchedBreakpoints } = useBreakpoint();
    const Component = as || "div";
    const gridItemColSpan = resolveResponsiveValue(colSpan, matchedBreakpoints);
    const gridItemRowSpan = resolveResponsiveValue(rowSpan, matchedBreakpoints);
    const gridItemMargin = resolveResponsiveValue(margin, matchedBreakpoints);
    const gridItemPadding = resolveResponsiveValue(padding, matchedBreakpoints);
    const gridColumnStart = gridItemColSpan ? `span ${gridItemColSpan}` : colStart;
    const gridColumnEnd = gridItemColSpan ? `span ${gridItemColSpan}` : colEnd;
    const gridRowStart = gridItemRowSpan ? `span ${gridItemRowSpan}` : rowStart;
    const gridRowEnd = gridItemRowSpan ? `span ${gridItemRowSpan}` : rowEnd;
    const gridStyles = {
      "--gridItem-margin": parseSpacing(gridItemMargin),
      "--gridItem-padding": parseSpacing(gridItemPadding),
      ...style,
      "--gridItem-justifySelf": horizontalAlignment,
      "--gridItem-alignSelf": verticalAlignment,
      "--gridItem-gridRowStart": gridRowStart,
      "--gridItem-gridColumnStart": gridColumnStart,
      "--gridItem-gridRowEnd": gridRowEnd,
      "--gridItem-gridColumnEnd": gridColumnEnd
    };
    return /* @__PURE__ */ jsx(
      Component,
      {
        className: clsx(withBaseName(), className),
        style: gridStyles,
        ref,
        ...rest,
        children
      }
    );
  }
);

export { GRID_ALIGNMENT_BASE, GridItem };
