// @salt-ds/core/dist-cjs/toggle-button-group/ToggleButtonGroup.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
var useControlled = require('../utils/useControlled.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var ToggleButtonGroup$1 = require('./ToggleButtonGroup.css.js');
var ToggleButtonGroupContext = require('./ToggleButtonGroupContext.js');

const withBaseName = makePrefixer.makePrefixer("saltToggleButtonGroup");
const ToggleButtonGroup = React.forwardRef(function ToggleButtonGroup2(props, ref) {
  const {
    appearance,
    children,
    className,
    value: valueProp,
    defaultValue,
    disabled,
    onChange,
    onKeyDown,
    orientation = "horizontal",
    readOnly,
    sentiment,
    ...rest
  } = props;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-toggle-button-group",
    css: ToggleButtonGroup$1,
    window: targetWindow
  });
  const groupRef = React.useRef(null);
  const handleRef = useForkRef.useForkRef(ref, groupRef);
  const [value, setValue] = useControlled.useControlled({
    default: defaultValue,
    controlled: valueProp,
    name: "ToggleButtonGroup",
    state: "value"
  });
  const [focused, setFocused] = React.useState(value);
  const select = React.useCallback(
    (event) => {
      const newValue = event.currentTarget.value;
      setValue(newValue);
      if (value !== newValue) {
        onChange == null ? void 0 : onChange(event);
      }
    },
    [onChange, value]
  );
  const isSelected = React.useCallback(
    (id) => {
      return value === id;
    },
    [value]
  );
  const focus = React.useCallback((id) => {
    setFocused(id);
  }, []);
  const isFocused = React.useCallback(
    (id) => {
      return focused === id || !focused;
    },
    [focused]
  );
  const contextValue = React.useMemo(
    () => ({
      appearance,
      disabled,
      focus,
      isFocused,
      isSelected,
      orientation,
      readOnly,
      select,
      sentiment
    }),
    [
      appearance,
      disabled,
      focus,
      isFocused,
      isSelected,
      orientation,
      readOnly,
      select,
      sentiment
    ]
  );
  const handleKeyDown = (event) => {
    var _a, _b, _c;
    const elements = Array.from(
      ((_a = groupRef.current) == null ? void 0 : _a.querySelectorAll("button:not([disabled])")) ?? []
    );
    const currentIndex = elements.findIndex(
      (element) => element === document.activeElement
    );
    switch (event.key) {
      case "ArrowDown":
      case "ArrowRight":
        (_b = elements[(currentIndex + 1) % elements.length]) == null ? void 0 : _b.focus();
        break;
      case "ArrowUp":
      case "ArrowLeft":
        (_c = elements[(currentIndex - 1 + elements.length) % elements.length]) == null ? void 0 : _c.focus();
        break;
    }
    onKeyDown == null ? void 0 : onKeyDown(event);
  };
  return /* @__PURE__ */ jsxRuntime.jsx(ToggleButtonGroupContext.ToggleButtonGroupContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      "aria-disabled": disabled,
      "aria-readonly": readOnly,
      className: clsx.clsx(
        withBaseName(),
        withBaseName(orientation),
        disabled && withBaseName("disabled"),
        readOnly && withBaseName("readOnly"),
        className
      ),
      role: "radiogroup",
      ref: handleRef,
      onKeyDown: handleKeyDown,
      ...rest,
      children
    }
  ) });
});

exports.ToggleButtonGroup = ToggleButtonGroup;
