// @salt-ds/core/dist-es/option/OptionListBase.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './OptionList.css.js';

const withBaseName = makePrefixer("saltOptionList");
const OptionListBase = forwardRef(
  function OptionListBase2(props, ref) {
    const { children, className, ...rest } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-option-list",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(withBaseName("container"), className),
        ...rest,
        ref,
        children
      }
    );
  }
);

export { OptionListBase };
