// @salt-ds/core/dist-cjs/pagination/Paginator.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var Button = require('../button/Button.js');
var SemanticIconProvider = require('../semantic-icon-provider/SemanticIconProvider.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var PageRanges = require('./PageRanges.js');
var usePaginationContext = require('./usePaginationContext.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Paginator$1 = require('./Paginator.css.js');

const withBaseName = makePrefixer.makePrefixer("saltPaginator");
const Paginator = React.forwardRef(
  function Paginator2({ className, boundaryCount, siblingCount, ...restProps }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-paginator",
      css: Paginator$1,
      window: targetWindow
    });
    const { count, page, onPageChange } = usePaginationContext.usePaginationContext();
    const { NextIcon, PreviousIcon } = SemanticIconProvider.useIcon();
    const onPreviousPage = (event) => {
      onPageChange(event, Math.max(1, page - 1));
    };
    const onNextPage = (event) => {
      onPageChange(event, Math.min(page + 1, count));
    };
    const isOnFirstPage = page === 1;
    const isOnLastPage = page === count;
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: clsx.clsx(withBaseName(), className), ...restProps, ref, children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        Button.Button,
        {
          appearance: "transparent",
          "aria-label": "Previous Page",
          onClick: onPreviousPage,
          disabled: isOnFirstPage,
          className: withBaseName("arrowButton-previous"),
          children: /* @__PURE__ */ jsxRuntime.jsx(PreviousIcon, { "aria-hidden": true })
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(PageRanges.PageRanges, { siblingCount, boundaryCount }),
      /* @__PURE__ */ jsxRuntime.jsx(
        Button.Button,
        {
          appearance: "transparent",
          "aria-label": "Next Page",
          onClick: onNextPage,
          disabled: isOnLastPage,
          className: withBaseName("arrowButton-next"),
          children: /* @__PURE__ */ jsxRuntime.jsx(NextIcon, { "aria-hidden": true })
        }
      )
    ] });
  }
);

exports.Paginator = Paginator;
