// @salt-ds/core/dist-cjs/stepper/Stepper.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Stepper$1 = require('./Stepper.css.js');
var StepperProvider = require('./internal/StepperProvider.js');

const withBaseName = makePrefixer.makePrefixer("saltStepper");
const Stepper = React.forwardRef(
  function Stepper2({ orientation: orientationProp, children, className, ...props }, ref) {
    const targetWindow = window.useWindow();
    const orientationContext = React.useContext(StepperProvider.StepperOrientationContext);
    const orientation = orientationProp || orientationContext;
    styles.useComponentCssInjection({
      testId: "salt-Stepper",
      css: Stepper$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(StepperProvider.StepperProvider, { orientation, children: /* @__PURE__ */ jsxRuntime.jsx(
      "ol",
      {
        className: clsx.clsx(withBaseName(), withBaseName(orientation), className),
        ref,
        ...props,
        children
      }
    ) });
  }
);

exports.Stepper = Stepper;
