// @salt-ds/core/dist-cjs/stepper/Step.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
var useControlled = require('../utils/useControlled.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useId = require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Step$1 = require('./Step.css.js');
var Stepper = require('./Stepper.js');
var StepConnector = require('./internal/StepConnector.js');
var StepExpandTrigger = require('./internal/StepExpandTrigger.js');
var StepIcon = require('./internal/StepIcon.js');
var StepScreenReaderOnly = require('./internal/StepScreenReaderOnly.js');
var StepText = require('./internal/StepText.js');
var StepperProvider = require('./internal/StepperProvider.js');

const withBaseName = makePrefixer.makePrefixer("saltStep");
const Step = React.forwardRef(function Step2({
  id: idProp,
  label,
  description,
  status,
  stage = "pending",
  expanded: expandedProp,
  defaultExpanded,
  onToggle,
  className,
  style,
  children,
  ...rest
}, ref) {
  const id = useId.useId(idProp);
  const targetWindow = window.useWindow();
  const depth = React.useContext(StepperProvider.StepDepthContext);
  const orientation = React.useContext(StepperProvider.StepperOrientationContext);
  const hasNestedSteps = !!children;
  const [expanded, setExpanded] = useControlled.useControlled({
    name: "Step",
    state: "expanded",
    controlled: expandedProp,
    default: Boolean(defaultExpanded)
  });
  styles.useComponentCssInjection({
    testId: "salt-step",
    css: Step$1,
    window: targetWindow
  });
  React.useEffect(() => {
    if (process.env.NODE_ENV !== "production") {
      if (depth === -1) {
        console.warn("<Step /> should be used within a <Stepper /> component!");
      }
      if (depth > 2) {
        console.warn("<Step /> should not be nested more than 2 levels deep!");
      }
      if (orientation === "horizontal" && hasNestedSteps) {
        console.warn(
          "<Stepper /> does not support nested steps in horizontal orientation!"
        );
      }
    }
  }, [depth, orientation]);
  const ariaCurrent = stage === "active" ? "step" : void 0;
  const iconSizeMultiplier = depth === 0 ? 1.5 : 1;
  const stageText = stage === "inprogress" ? "in progress" : stage;
  const state = status || stageText;
  const labelId = `${id}-label`;
  const descriptionId = `${id}-description`;
  const expandTriggerId = `${id}-expand-trigger`;
  const nestedStepperId = `${id}-nested-stepper`;
  const screenReaderOnly = {
    stateId: `${id}-sr-only-state`,
    stateText: state !== "active" ? state : "",
    substepsId: `${id}-sr-only-substeps`,
    substepsText: "substeps",
    toggleSubstepsId: `${id}-sr-only-toggle-substeps`,
    toggleSubstepsText: "toggle substeps"
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "li",
    {
      id,
      "aria-current": ariaCurrent,
      className: clsx(
        withBaseName(),
        withBaseName(`stage-${stage}`),
        withBaseName(`depth-${depth}`),
        status && withBaseName(`status-${status}`),
        !hasNestedSteps && withBaseName("terminal"),
        hasNestedSteps && expanded && withBaseName("expanded"),
        hasNestedSteps && !expanded && withBaseName("collapsed"),
        className
      ),
      style: {
        "--saltStep-depth": depth,
        ...style
      },
      ...rest,
      ref,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(StepScreenReaderOnly.StepScreenReaderOnly, { children: `${label} ${description !== void 0 ? description : ""} ${screenReaderOnly.stateText}` }),
        hasNestedSteps && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            StepScreenReaderOnly.StepScreenReaderOnly,
            {
              id: screenReaderOnly.toggleSubstepsId,
              "aria-hidden": true,
              children: screenReaderOnly.toggleSubstepsText
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(StepScreenReaderOnly.StepScreenReaderOnly, { id: screenReaderOnly.substepsId, "aria-hidden": true, children: screenReaderOnly.substepsText }),
          /* @__PURE__ */ jsxRuntime.jsx(StepScreenReaderOnly.StepScreenReaderOnly, { id: screenReaderOnly.stateId, "aria-hidden": true, children: screenReaderOnly.stateText })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsx(StepConnector.StepConnector, {}),
        /* @__PURE__ */ jsxRuntime.jsx(
          StepIcon.StepIcon,
          {
            stage,
            status,
            sizeMultiplier: iconSizeMultiplier,
            "aria-hidden": true
          }
        ),
        label && /* @__PURE__ */ jsxRuntime.jsx(StepText.StepText, { id: labelId, purpose: "label", "aria-hidden": true, children: label }),
        description && /* @__PURE__ */ jsxRuntime.jsx(StepText.StepText, { id: descriptionId, purpose: "description", "aria-hidden": true, children: description }),
        hasNestedSteps && orientation === "vertical" && /* @__PURE__ */ jsxRuntime.jsx(
          StepExpandTrigger.StepExpandTrigger,
          {
            id: expandTriggerId,
            "aria-expanded": expanded,
            "aria-controls": nestedStepperId,
            "aria-labelledby": [
              labelId,
              descriptionId,
              screenReaderOnly.stateId,
              screenReaderOnly.toggleSubstepsId
            ].join(" "),
            expanded,
            onClick: (event) => {
              onToggle == null ? void 0 : onToggle(event);
              setExpanded(!expanded);
            }
          }
        ),
        hasNestedSteps && orientation === "vertical" && /* @__PURE__ */ jsxRuntime.jsx(
          Stepper.Stepper,
          {
            id: nestedStepperId,
            "aria-labelledby": [labelId, screenReaderOnly.substepsId].join(" "),
            "aria-hidden": !expanded,
            hidden: !expanded,
            children
          }
        )
      ]
    }
  );
});

exports.Step = Step;
