// @salt-ds/core/dist-cjs/panel/Panel.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Panel$1 = require('./Panel.css.js');

const withBaseName = makePrefixer.makePrefixer("saltPanel");
const Panel = React.forwardRef(function Panel2({ className, children, variant = "primary", ...restProps }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-panel",
    css: Panel$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      className: clsx.clsx(withBaseName(), withBaseName(variant), className),
      ref,
      ...restProps,
      children
    }
  );
});

exports.Panel = Panel;
