// @salt-ds/core/dist-es/file-drop-zone/FileDropZoneTrigger.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { forwardRef, useRef } from 'react';
import { Button } from '../button/Button.js';
import 'clsx';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';

const FileDropZoneTrigger = forwardRef(function FileDropZoneTrigger2({ accept, children, disabled, multiple = false, onChange, ...rest }, ref) {
  const buttonRef = useRef(null);
  const fileInputRef = useRef(null);
  const triggerRef = useForkRef(ref, buttonRef);
  const handleFocus = (event) => {
    var _a;
    event.stopPropagation();
    (_a = buttonRef.current) == null ? void 0 : _a.focus();
  };
  const handleClick = (event) => {
    var _a;
    event.stopPropagation();
    (_a = fileInputRef.current) == null ? void 0 : _a.click();
  };
  const handleChange = (event) => {
    const files = Array.from(event.target.files ?? []);
    event.target.value = "";
    onChange == null ? void 0 : onChange(event, files);
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Button,
      {
        onClick: handleClick,
        disabled,
        ref: triggerRef,
        ...rest,
        children: children ?? "Browse files"
      }
    ),
    /* @__PURE__ */ jsx(
      "input",
      {
        accept,
        className: "input-hidden",
        disabled,
        multiple,
        onChange: handleChange,
        onFocus: handleFocus,
        ref: fileInputRef,
        type: "file"
      }
    )
  ] });
});

export { FileDropZoneTrigger };
