// @salt-ds/core/dist-es/card/Card.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { capitalize } from '../utils/capitalize.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Card.css.js';

const withBaseName = makePrefixer("saltCard");
const Card = forwardRef(
  function Card2(props, ref) {
    const {
      accent,
      children,
      className,
      disabled,
      interactable,
      hoverable,
      variant = "primary",
      ...rest
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-card",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("accent")]: accent,
            [withBaseName(`accent${capitalize(accent || "")}`)]: accent,
            [withBaseName("hoverable")]: hoverable,
            /* **Deprecated:** InteractableCard should be used instead for these features */
            [withBaseName("disabled")]: disabled,
            [withBaseName("interactable")]: interactable
          },
          className
        ),
        ref,
        ...rest,
        children
      }
    );
  }
);

export { Card };
