// @salt-ds/core/dist-es/interactable-card/InteractableCard.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useRef } from 'react';
import { capitalize } from '../utils/capitalize.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useControlled } from '../utils/useControlled.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './InteractableCard.css.js';
import { useInteractableCardGroup } from './InteractableCardGroupContext.js';
import { useInteractableCard } from './useInteractableCard.js';

const withBaseName = makePrefixer("saltInteractableCard");
const InteractableCard = forwardRef(function InteractableCard2(props, ref) {
  const {
    accent,
    accentPlacement,
    children,
    className,
    disabled: disabledProp,
    variant = "primary",
    value,
    onBlur,
    onChange,
    onClick,
    onKeyUp,
    onKeyDown,
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-interactable-card",
    css: css_248z,
    window: targetWindow
  });
  const interactableCardGroup = useInteractableCardGroup();
  const interactableCardGroupSelected = interactableCardGroup == null ? void 0 : interactableCardGroup.isSelected(value);
  const disabled = (interactableCardGroup == null ? void 0 : interactableCardGroup.disabled) || disabledProp;
  const [selected, setSelected] = useControlled({
    controlled: interactableCardGroupSelected,
    default: Boolean(false),
    name: "InteractableCard",
    state: "selected"
  });
  const role = interactableCardGroup ? interactableCardGroup.multiSelect ? "checkbox" : "radio" : "button";
  const isMultiselect = interactableCardGroup == null ? void 0 : interactableCardGroup.multiSelect;
  const isFirstChild = interactableCardGroup == null ? void 0 : interactableCardGroup.isFirstChild(value);
  const ariaChecked = role === "radio" || role === "checkbox" ? selected : void 0;
  const accentValue = accent || accentPlacement;
  const handleClick = (event) => {
    if (interactableCardGroup && !disabled) {
      interactableCardGroup.select(event, value);
      setSelected(!selected);
    }
    onChange == null ? void 0 : onChange(event);
    onClick == null ? void 0 : onClick(event);
  };
  let tabIndex;
  if (interactableCardGroup) {
    if (disabled) {
      tabIndex = -1;
    } else if (isMultiselect) {
      tabIndex = 0;
    } else {
      tabIndex = selected ? 0 : -1;
      if (!interactableCardGroup.value && isFirstChild) {
        tabIndex = 0;
      }
    }
  } else {
    tabIndex = disabled ? -1 : 0;
  }
  const cardRef = useRef(null);
  const handleRef = useForkRef(ref, cardRef);
  const { active, cardProps } = useInteractableCard({
    disabled,
    onKeyUp,
    onKeyDown,
    onBlur,
    onClick
  });
  return /* @__PURE__ */ jsx(
    "div",
    {
      ...cardProps,
      role,
      "aria-checked": ariaChecked,
      "aria-disabled": disabled,
      "data-value": value,
      className: clsx(
        withBaseName(),
        withBaseName(variant),
        {
          [withBaseName("accent")]: accentValue,
          [withBaseName(`accent${capitalize(accentValue ?? "")}`)]: accentValue,
          [withBaseName("active")]: role === "button" && active,
          [withBaseName("disabled")]: disabled,
          [withBaseName("selected")]: selected
        },
        className
      ),
      ...rest,
      onClick: handleClick,
      ref: handleRef,
      tabIndex,
      children
    }
  );
});

export { InteractableCard };
