// @salt-ds/core/dist-es/interactable-card/InteractableCardGroup.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useRef, useState, useEffect, useCallback, useMemo } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useControlled } from '../utils/useControlled.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './InteractableCardGroup.css.js';
import { InteractableCardGroupContext } from './InteractableCardGroupContext.js';

const withBaseName = makePrefixer("saltInteractableCardGroup");
const InteractableCardGroup = forwardRef(function InteractableCardGroup2(props, ref) {
  const {
    children,
    className,
    value: valueProp,
    defaultValue,
    disabled,
    onChange,
    onKeyDown,
    multiSelect,
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-interactable-card-group",
    css: css_248z,
    window: targetWindow
  });
  const groupRef = useRef(null);
  const handleRef = useForkRef(ref, groupRef);
  const [value, setValue] = useControlled({
    default: defaultValue,
    controlled: valueProp,
    name: "InteractableCardGroup",
    state: "value"
  });
  const [elements, setElements] = useState([]);
  useEffect(() => {
    var _a;
    const childElements = Array.from(
      ((_a = groupRef.current) == null ? void 0 : _a.querySelectorAll(
        ".saltInteractableCard:not([disabled])"
      )) ?? []
    );
    setElements(childElements);
  }, [children]);
  const select = useCallback(
    (event, newValue) => {
      if (multiSelect) {
        const currentValues = Array.isArray(value) ? value : [];
        const isSelected2 = currentValues.includes(newValue);
        const nextValues = isSelected2 ? currentValues.filter((val) => val !== newValue) : [...currentValues, newValue];
        setValue(nextValues);
        onChange == null ? void 0 : onChange(event, nextValues);
      } else {
        setValue(newValue);
        if (value !== newValue) {
          onChange == null ? void 0 : onChange(event, newValue);
        }
      }
    },
    [onChange, value, multiSelect]
  );
  const isSelected = useCallback(
    (cardValue) => multiSelect ? Array.isArray(value) && value.includes(cardValue) : cardValue !== void 0 && value === cardValue,
    [value, multiSelect]
  );
  const isFirstChild = useCallback(
    (cardValue) => {
      return elements.findIndex(
        (element) => element.getAttribute("data-value") === cardValue
      ) === 0;
    },
    [elements]
  );
  const contextValue = useMemo(
    () => ({
      select,
      isSelected,
      isFirstChild,
      disabled,
      multiSelect,
      value
    }),
    [select, isSelected, disabled, multiSelect, isFirstChild, value]
  );
  const handleKeyDown = (event) => {
    var _a, _b;
    const currentIndex = elements.findIndex(
      (element) => element === document.activeElement
    );
    const nextIndex = (currentIndex + 1) % elements.length;
    const prevIndex = (currentIndex - 1 + elements.length) % elements.length;
    if (event.key === " ") {
      event.preventDefault();
      select(
        event,
        elements[currentIndex].getAttribute(
          "data-value"
        )
      );
    }
    if (!multiSelect) {
      switch (event.key) {
        case "ArrowDown":
        case "ArrowRight":
          select(
            event,
            elements[nextIndex].getAttribute(
              "data-value"
            )
          );
          (_a = elements[nextIndex]) == null ? void 0 : _a.focus();
          break;
        case "ArrowUp":
        case "ArrowLeft":
          select(
            event,
            elements[prevIndex].getAttribute(
              "data-value"
            )
          );
          (_b = elements[prevIndex]) == null ? void 0 : _b.focus();
          break;
      }
    }
    onKeyDown == null ? void 0 : onKeyDown(event);
  };
  return /* @__PURE__ */ jsx(InteractableCardGroupContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsx(
    "div",
    {
      className: clsx(withBaseName(), className),
      role: multiSelect ? "group" : "radiogroup",
      onKeyDown: handleKeyDown,
      ref: handleRef,
      ...rest,
      children
    }
  ) });
});

export { InteractableCardGroup };
