// @salt-ds/core/dist-cjs/form-field/FormField.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var FormFieldContext = require('../form-field-context/FormFieldContext.js');
var capitalize = require('../utils/capitalize.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useId = require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var FormField$1 = require('./FormField.css.js');

const withBaseName = makePrefixer.makePrefixer("saltFormField");
const FormField = React.forwardRef(
  ({
    children,
    className,
    disabled = false,
    id: idProp,
    labelPlacement = "top",
    necessity,
    readOnly = false,
    validationStatus,
    ...restProps
  }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-form-field",
      css: FormField$1,
      window: targetWindow
    });
    const formId = useId.useId(idProp);
    const labelId = formId ? `label-${formId}` : void 0;
    const helperTextId = formId ? `helperText-${formId}` : void 0;
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        ref,
        className: clsx.clsx(
          withBaseName(),
          {
            [withBaseName("disabled")]: disabled,
            [withBaseName(`label${capitalize.capitalize(labelPlacement)}`)]: labelPlacement
          },
          className
        ),
        ...restProps,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          FormFieldContext.FormFieldContext.Provider,
          {
            value: {
              a11yProps: {
                "aria-labelledby": labelId,
                "aria-describedby": helperTextId
              },
              disabled,
              necessity,
              readOnly,
              validationStatus
            },
            children
          }
        )
      }
    );
  }
);

exports.FormField = FormField;
