// @salt-ds/core/dist-cjs/form-field/FormFieldLabel.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
require('../form-field-context/FormFieldContext.js');
var useFormFieldProps = require('../form-field-context/useFormFieldProps.js');
require('../text/Text.js');
require('../text/Code.js');
require('../text/Display.js');
require('../text/Headings.js');
var Label = require('../text/Label.js');
require('../text/TextAction.js');
require('../text/TextNotation.js');
var capitalize = require('../utils/capitalize.js');
require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var FormFieldLabel$1 = require('./FormFieldLabel.css.js');

const withBaseName = makePrefixer.makePrefixer("saltFormFieldLabel");
const FormFieldLabel = ({
  className,
  children,
  intent = "label",
  ...restProps
}) => {
  const { a11yProps, disabled, necessity } = useFormFieldProps.useFormFieldProps();
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-form-field-label",
    css: FormFieldLabel$1,
    window: targetWindow
  });
  const necessityLabel = necessity ? necessity === "asterisk" ? " *" : ` (${capitalize.capitalize(necessity)})` : void 0;
  return /* @__PURE__ */ jsxRuntime.jsxs(
    Label.Label,
    {
      as: "label",
      className: clsx.clsx(withBaseName(), withBaseName(intent), className),
      id: a11yProps == null ? void 0 : a11yProps["aria-labelledby"],
      disabled,
      variant: "secondary",
      ...restProps,
      children: [
        children,
        necessityLabel && /* @__PURE__ */ jsxRuntime.jsx("span", { className: withBaseName("necessityLabel"), children: necessityLabel })
      ]
    }
  );
};

exports.FormFieldLabel = FormFieldLabel;
