// @salt-ds/core/dist-es/spinner/Spinner.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef, useEffect } from 'react';
import '../aria-announcer/AriaAnnouncerContext.js';
import { useAriaAnnouncer } from '../aria-announcer/useAriaAnnouncer.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useId } from '../utils/useId.js';
import { useDensity } from '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { SpinnerSVG } from './svgSpinners/SpinnerSVG.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './Spinner.css.js';

const SpinnerSizeValues = [
  "default",
  "large",
  "small",
  "medium"
];
const handleSize = (size) => size === "default" ? "medium" : size;
const withBaseName = makePrefixer("saltSpinner");
const Spinner = forwardRef(
  function Spinner2({
    "aria-label": ariaLabel = "loading",
    announcerInterval = 5e3,
    announcerTimeout = 2e4,
    completionAnnouncement = `finished ${ariaLabel}`,
    disableAnnouncer,
    role = "img",
    className,
    size = "medium",
    id: idProp,
    ...rest
  }, ref) {
    const id = useId(idProp);
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-spinner",
      css: css_248z,
      window: targetWindow
    });
    const { announce } = useAriaAnnouncer();
    const density = useDensity();
    size = handleSize(size);
    useEffect(() => {
      if (disableAnnouncer) return;
      announce(ariaLabel);
      const startTime = (/* @__PURE__ */ new Date()).getTime();
      const interval = announcerInterval > 0 && setInterval(() => {
        if ((/* @__PURE__ */ new Date()).getTime() - startTime > announcerTimeout) {
          announce(
            `${ariaLabel} is still in progress, but will no longer announce.`
          );
          interval && clearInterval(interval);
          return;
        }
        announce(ariaLabel);
      }, announcerInterval);
      return () => {
        if (disableAnnouncer) return;
        interval && clearInterval(interval);
        if (completionAnnouncement) {
          announce(completionAnnouncement);
        }
      };
    }, [
      announce,
      announcerInterval,
      announcerTimeout,
      ariaLabel,
      completionAnnouncement,
      disableAnnouncer
    ]);
    return /* @__PURE__ */ jsx(
      "div",
      {
        "aria-label": ariaLabel,
        className: clsx(withBaseName(), withBaseName(size), className),
        ref,
        role,
        ...rest,
        children: /* @__PURE__ */ jsx(SpinnerSVG, { size, density, id })
      }
    );
  }
);

export { Spinner, SpinnerSizeValues };
