// @salt-ds/core/dist-es/status-indicator/StatusIndicator.js
import { jsx } from 'react/jsx-runtime';
import { DEFAULT_ICON_SIZE } from '@salt-ds/icons';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { useIcon } from '../semantic-icon-provider/SemanticIconProvider.js';
import css_248z from './StatusIndicator.css.js';

const statusToAriaLabelMap = {
  error: "error",
  success: "success",
  warning: "warning",
  info: "info"
};
const withBaseName = makePrefixer("saltStatusIndicator");
const StatusIndicator = forwardRef(
  function StatusIndicator2({ className, status, size = DEFAULT_ICON_SIZE, ...restProps }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-status-indicator",
      css: css_248z,
      window: targetWindow
    });
    const { ErrorIcon, WarningIcon, SuccessIcon, InfoIcon } = useIcon();
    const iconMap = {
      error: ErrorIcon,
      success: SuccessIcon,
      warning: WarningIcon,
      info: InfoIcon
    };
    const IconComponent = iconMap[status];
    const ariaLabel = statusToAriaLabelMap[status];
    return /* @__PURE__ */ jsx(
      IconComponent,
      {
        className: clsx(withBaseName(), withBaseName(status), className),
        size,
        "aria-label": ariaLabel,
        ...restProps,
        ref
      }
    );
  }
);

export { StatusIndicator };
