// @salt-ds/core/dist-es/divider/Divider.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Divider.css.js';

const withBaseName = makePrefixer("saltDivider");
const Divider = forwardRef(
  function Divider2(props, ref) {
    const {
      className,
      orientation = "horizontal",
      variant = "primary",
      ...rest
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-divider",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(
          withBaseName(),
          withBaseName(orientation),
          withBaseName(variant),
          className
        ),
        "aria-orientation": orientation,
        role: "separator",
        ref,
        ...rest
      }
    );
  }
);

export { Divider };
