// @salt-ds/core/dist-cjs/drawer/Drawer.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('@floating-ui/react');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Scrim = require('../scrim/Scrim.js');
var makePrefixer = require('../utils/makePrefixer.js');
var useFloatingUI = require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Drawer$1 = require('./Drawer.css.js');

const ConditionalScrimWrapper = ({
  condition,
  children
}) => {
  return condition ? /* @__PURE__ */ jsxRuntime.jsxs(Scrim.Scrim, { fixed: true, children: [
    " ",
    children,
    " "
  ] }) : /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    children,
    " "
  ] });
};
const withBaseName = makePrefixer.makePrefixer("saltDrawer");
const Drawer = React.forwardRef(
  function Drawer2(props, ref) {
    var _a, _b;
    const {
      children,
      className,
      position = "left",
      open = false,
      onOpenChange,
      variant = "primary",
      disableDismiss,
      disableScrim,
      ...rest
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-drawer",
      css: Drawer$1,
      window: targetWindow
    });
    const [showComponent, setShowComponent] = React.useState(false);
    const { Component: FloatingComponent } = useFloatingUI.useFloatingComponent();
    const { context, floating, elements } = useFloatingUI.useFloatingUI({
      open: showComponent,
      onOpenChange
    });
    const { getFloatingProps } = react.useInteractions([
      react.useClick(context),
      react.useDismiss(context, { enabled: !disableDismiss })
    ]);
    const handleRef = useForkRef.useForkRef(floating, ref);
    React.useEffect(() => {
      if (open && !showComponent) {
        setShowComponent(true);
      }
      if (!open && showComponent) {
        const animate = setTimeout(() => {
          setShowComponent(false);
        }, 300);
        return () => clearTimeout(animate);
      }
    }, [open, showComponent]);
    return /* @__PURE__ */ jsxRuntime.jsx(ConditionalScrimWrapper, { condition: showComponent && !disableScrim, children: /* @__PURE__ */ jsxRuntime.jsx(
      FloatingComponent,
      {
        open: showComponent,
        ref: handleRef,
        role: "dialog",
        width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
        height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight,
        "aria-modal": "true",
        focusManagerProps: {
          context
        },
        className: clsx.clsx(
          withBaseName(),
          withBaseName(position),
          {
            [withBaseName("enterAnimation")]: open,
            [withBaseName("exitAnimation")]: !open,
            [withBaseName(variant)]: variant
          },
          className
        ),
        ...getFloatingProps(),
        ...rest,
        children
      }
    ) });
  }
);

exports.Drawer = Drawer;
