// @salt-ds/core/dist-cjs/drawer/DrawerCloseButton.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Button = require('../button/Button.js');
var SemanticIconProvider = require('../semantic-icon-provider/SemanticIconProvider.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var DrawerCloseButton$1 = require('./DrawerCloseButton.css.js');

const withBaseName = makePrefixer.makePrefixer("saltDrawerCloseButton");
const DrawerCloseButton = React.forwardRef(
  function DrawerCloseButton2({ className, ...rest }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-drawer-close-button",
      css: DrawerCloseButton$1,
      window: targetWindow
    });
    const { CloseIcon } = SemanticIconProvider.useIcon();
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("container"), children: /* @__PURE__ */ jsxRuntime.jsx(
      Button.Button,
      {
        ref,
        "aria-label": "Close Drawer",
        appearance: "transparent",
        className: clsx(withBaseName(), className),
        ...rest,
        children: /* @__PURE__ */ jsxRuntime.jsx(CloseIcon, { "aria-hidden": true })
      }
    ) });
  }
);

exports.DrawerCloseButton = DrawerCloseButton;
