// @salt-ds/core/dist-es/avatar/Avatar.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { useIcon } from '../semantic-icon-provider/SemanticIconProvider.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Avatar.css.js';
import { useAvatarImage } from './useAvatarImage.js';

const withBaseName = makePrefixer("saltAvatar");
const DEFAULT_AVATAR_SIZE = 2;
const defaultNameToInitials = (name) => name == null ? void 0 : name.split(" ").slice(0, 2).map((n) => n[0]).join("").toUpperCase();
const Avatar = forwardRef(function Avatar2({
  className,
  children: childrenProp,
  color = "accent",
  name,
  nameToInitials = defaultNameToInitials,
  src,
  size = DEFAULT_AVATAR_SIZE,
  style: styleProp,
  fallbackIcon: fallbackIconProp,
  ...rest
}, ref) {
  const targetWindow = useWindow();
  const { UserIcon } = useIcon();
  const fallbackIcon = fallbackIconProp === void 0 ? /* @__PURE__ */ jsx(UserIcon, { "aria-hidden": true }) : fallbackIconProp;
  useComponentCssInjection({
    testId: "salt-avatar",
    css: css_248z,
    window: targetWindow
  });
  let children;
  const style = {
    ...styleProp,
    "--saltAvatar-size-multiplier": `${size}`
  };
  const status = useAvatarImage({ src });
  const hasImgNotFailing = status === "loaded";
  if (hasImgNotFailing) {
    children = /* @__PURE__ */ jsx("img", { alt: "", src });
  } else if (childrenProp != null) {
    children = childrenProp;
  }
  const avatarInitials = nameToInitials(name);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      style,
      className: clsx(
        withBaseName(),
        withBaseName(color),
        {
          [withBaseName("withImage")]: hasImgNotFailing
        },
        className
      ),
      role: name ? "img" : void 0,
      "aria-label": name,
      ...rest,
      children: children || avatarInitials || fallbackIcon
    }
  );
});

export { Avatar };
