// @salt-ds/core/dist-es/split-layout/SplitLayout.js
import { jsxs } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { FlexLayout } from '../flex-layout/FlexLayout.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './SplitLayout.css.js';

const withBaseName = makePrefixer("saltSplitLayout");
const SplitLayout = forwardRef(
  ({ endItem, startItem, className, ...rest }, ref) => {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-split-layout",
      css: css_248z,
      window: targetWindow
    });
    const justify = endItem && !startItem ? "end" : "space-between";
    return /* @__PURE__ */ jsxs(
      FlexLayout,
      {
        className: clsx(withBaseName(), className),
        ref,
        justify,
        ...rest,
        children: [
          startItem,
          endItem
        ]
      }
    );
  }
);

export { SplitLayout };
