// @salt-ds/core/dist-cjs/stepper/internal/StepExpandTrigger.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var Button = require('../../button/Button.js');
require('react');
var SemanticIconProvider = require('../../semantic-icon-provider/SemanticIconProvider.js');
var makePrefixer = require('../../utils/makePrefixer.js');
require('../../utils/useFloatingUI/useFloatingUI.js');
require('../../utils/useId.js');
require('../../salt-provider/SaltProvider.js');
require('../../viewport/ViewportProvider.js');
var StepExpandTrigger$1 = require('./StepExpandTrigger.css.js');

const withBaseName = makePrefixer.makePrefixer("saltStepExpandTrigger");
function StepExpandTrigger({
  id,
  expanded,
  className,
  ...props
}) {
  const { CollapseIcon, ExpandIcon } = SemanticIconProvider.useIcon();
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-step-expand-trigger",
    css: StepExpandTrigger$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsx(
    Button.Button,
    {
      id,
      sentiment: "neutral",
      appearance: "transparent",
      className: clsx(withBaseName(), className),
      ...props,
      children: expanded ? /* @__PURE__ */ jsxRuntime.jsx(CollapseIcon, { "aria-hidden": true }) : /* @__PURE__ */ jsxRuntime.jsx(ExpandIcon, { "aria-hidden": true })
    }
  );
}

exports.StepExpandTrigger = StepExpandTrigger;
