// @salt-ds/core/dist-es/overlay/OverlayPanelCloseButton.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { forwardRef } from 'react';
import { Button } from '../button/Button.js';
import { useIcon } from '../semantic-icon-provider/SemanticIconProvider.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './OverlayPanelCloseButton.css.js';

const withBaseName = makePrefixer("saltOverlayPanelCloseButton");
const OverlayPanelCloseButton = forwardRef(function OverlayPanelButton({ className, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-overlay-panel-close-button",
    css: css_248z,
    window: targetWindow
  });
  const { CloseIcon } = useIcon();
  return /* @__PURE__ */ jsx("div", { className: clsx(withBaseName(), className), children: /* @__PURE__ */ jsx(
    Button,
    {
      ref,
      "aria-label": "Close overlay",
      appearance: "transparent",
      className: withBaseName("button"),
      ...rest,
      children: /* @__PURE__ */ jsx(CloseIcon, { "aria-hidden": true })
    }
  ) });
});

export { OverlayPanelCloseButton };
