// @salt-ds/core/dist-es/overlay/Overlay.js
import { jsx } from 'react/jsx-runtime';
import { offset, flip, shift, arrow, limitShift, useInteractions, useRole, useClick, useDismiss } from '@floating-ui/react';
import { useRef, useMemo } from 'react';
import 'clsx';
import { useControlled } from '../utils/useControlled.js';
import { useFloatingUI } from '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { OverlayContext } from './OverlayContext.js';

const Overlay = ({
  children,
  open,
  onOpenChange,
  placement: placementProp = "top"
}) => {
  const arrowRef = useRef(null);
  const [openState, setOpenState] = useControlled({
    controlled: open,
    default: false,
    name: "Overlay",
    state: "open"
  });
  const handleOpenChange = (newOpen) => {
    setOpenState(newOpen);
    onOpenChange == null ? void 0 : onOpenChange(newOpen);
  };
  const { x, y, strategy, context, elements, floating, reference } = useFloatingUI({
    open: openState,
    onOpenChange: handleOpenChange,
    placement: placementProp,
    middleware: [
      offset(11),
      flip(),
      shift({ limiter: limitShift() }),
      arrow({ element: arrowRef })
    ]
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useRole(context, { role: "dialog" }),
    useClick(context),
    useDismiss(context)
  ]);
  const floatingStyles = useMemo(() => {
    var _a, _b;
    return {
      top: y ?? 0,
      left: x ?? 0,
      position: strategy,
      width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
      height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight
    };
  }, [elements.floating, strategy, x, y]);
  const arrowProps = {
    ref: arrowRef,
    context
  };
  return /* @__PURE__ */ jsx(
    OverlayContext.Provider,
    {
      value: {
        openState,
        floatingStyles,
        context,
        arrowProps,
        floating,
        reference,
        getFloatingProps,
        getReferenceProps
      },
      children
    }
  );
};

export { Overlay };
