// @salt-ds/core/dist-cjs/split-layout/SplitLayout.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var FlexLayout = require('../flex-layout/FlexLayout.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var SplitLayout$1 = require('./SplitLayout.css.js');

const withBaseName = makePrefixer.makePrefixer("saltSplitLayout");
const SplitLayout = React.forwardRef(
  ({ endItem, startItem, className, ...rest }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-split-layout",
      css: SplitLayout$1,
      window: targetWindow
    });
    const justify = endItem && !startItem ? "end" : "space-between";
    return /* @__PURE__ */ jsxRuntime.jsxs(
      FlexLayout.FlexLayout,
      {
        className: clsx.clsx(withBaseName(), className),
        ref,
        justify,
        ...rest,
        children: [
          startItem,
          endItem
        ]
      }
    );
  }
);

exports.SplitLayout = SplitLayout;
