// @salt-ds/core/dist-cjs/checkbox/Checkbox.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
require('../form-field-context/FormFieldContext.js');
var useFormFieldProps = require('../form-field-context/useFormFieldProps.js');
var makePrefixer = require('../utils/makePrefixer.js');
var useControlled = require('../utils/useControlled.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
require('../utils/useId.js');
var useIsomorphicLayoutEffect = require('../utils/useIsomorphicLayoutEffect.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var CheckboxIcon = require('./CheckboxIcon.js');
var useCheckboxGroup = require('./internal/useCheckboxGroup.js');
var Checkbox$1 = require('./Checkbox.css.js');

const withBaseName = makePrefixer.makePrefixer("saltCheckbox");
const Checkbox = React.forwardRef(
  function Checkbox2({
    checked: checkedProp,
    className,
    defaultChecked,
    disabled: disabledProp,
    error,
    indeterminate,
    inputProps = {},
    inputRef: inputRefProp,
    label,
    name,
    onBlur,
    onChange,
    onFocus,
    value,
    validationStatus: validationStatusProp,
    readOnly: readOnlyProp,
    ...rest
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-checkbox",
      css: Checkbox$1,
      window: targetWindow
    });
    const checkboxGroup = useCheckboxGroup.useCheckboxGroup();
    const {
      "aria-describedby": inputDescribedBy,
      "aria-labelledby": inputLabelledBy,
      className: inputClassName,
      onChange: inputOnChange,
      ...restInputProps
    } = inputProps;
    const checkboxGroupChecked = checkedProp ?? ((checkboxGroup == null ? void 0 : checkboxGroup.checkedValues) && value ? checkboxGroup.checkedValues.includes(value) : checkedProp);
    const [checked, setChecked] = useControlled.useControlled({
      controlled: checkboxGroupChecked,
      default: Boolean(defaultChecked),
      name: "Checkbox",
      state: "checked"
    });
    const {
      a11yProps: formFieldA11yProps,
      disabled: formFieldDisabled,
      readOnly: formFieldReadOnly,
      validationStatus: formFieldValidationStatus
    } = useFormFieldProps.useFormFieldProps();
    const disabled = (checkboxGroup == null ? void 0 : checkboxGroup.disabled) || formFieldDisabled || disabledProp;
    const readOnly = (checkboxGroup == null ? void 0 : checkboxGroup.readOnly) || formFieldReadOnly || readOnlyProp;
    const validationStatus = !disabled ? (checkboxGroup == null ? void 0 : checkboxGroup.validationStatus) ?? formFieldValidationStatus ?? validationStatusProp : void 0;
    const inputRef = React.useRef(null);
    const handleInputRef = useForkRef.useForkRef(inputRefProp, inputRef);
    const handleChange = (event) => {
      var _a;
      if (event.nativeEvent.defaultPrevented || readOnly) {
        return;
      }
      const value2 = event.target.checked;
      setChecked(value2);
      onChange == null ? void 0 : onChange(event);
      inputOnChange == null ? void 0 : inputOnChange(event);
      (_a = checkboxGroup == null ? void 0 : checkboxGroup.onChange) == null ? void 0 : _a.call(checkboxGroup, event);
    };
    useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(() => {
      if (inputRef.current != null) {
        inputRef.current.indeterminate = indeterminate ?? false;
      }
    }, [indeterminate]);
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "label",
      {
        className: clsx.clsx(
          withBaseName(),
          {
            [withBaseName("disabled")]: disabled,
            [withBaseName("readOnly")]: readOnly,
            [withBaseName("error")]: error,
            [withBaseName(validationStatus ?? "")]: validationStatus
          },
          className
        ),
        ref,
        ...rest,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(
            "input",
            {
              "aria-readonly": readOnly || void 0,
              "aria-describedby": clsx.clsx(
                checkboxGroup === void 0 ? formFieldA11yProps == null ? void 0 : formFieldA11yProps["aria-describedby"] : void 0,
                inputDescribedBy
              ),
              "aria-labelledby": clsx.clsx(
                checkboxGroup === void 0 ? formFieldA11yProps == null ? void 0 : formFieldA11yProps["aria-labelledby"] : void 0,
                inputLabelledBy
              ),
              name,
              value,
              checked,
              className: clsx.clsx(withBaseName("input"), inputClassName),
              "data-indeterminate": indeterminate,
              defaultChecked,
              disabled,
              readOnly,
              onBlur,
              onChange: handleChange,
              onFocus,
              type: "checkbox",
              ref: handleInputRef,
              ...restInputProps
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsx(
            CheckboxIcon.CheckboxIcon,
            {
              checked,
              disabled,
              readOnly,
              indeterminate,
              validationStatus,
              error
            }
          ),
          label
        ]
      }
    );
  }
);

exports.Checkbox = Checkbox;
