// @salt-ds/core/dist-cjs/overlay/OverlayPanelCloseButton.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Button = require('../button/Button.js');
var SemanticIconProvider = require('../semantic-icon-provider/SemanticIconProvider.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var OverlayPanelCloseButton$1 = require('./OverlayPanelCloseButton.css.js');

const withBaseName = makePrefixer.makePrefixer("saltOverlayPanelCloseButton");
const OverlayPanelCloseButton = React.forwardRef(function OverlayPanelButton({ className, ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-overlay-panel-close-button",
    css: OverlayPanelCloseButton$1,
    window: targetWindow
  });
  const { CloseIcon } = SemanticIconProvider.useIcon();
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: clsx(withBaseName(), className), children: /* @__PURE__ */ jsxRuntime.jsx(
    Button.Button,
    {
      ref,
      "aria-label": "Close overlay",
      appearance: "transparent",
      className: withBaseName("button"),
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx(CloseIcon, { "aria-hidden": true })
    }
  ) });
});

exports.OverlayPanelCloseButton = OverlayPanelCloseButton;
