// @salt-ds/core/dist-es/dialog/Dialog.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useInteractions, useClick, useDismiss } from '@floating-ui/react';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useState, useEffect, useMemo } from 'react';
import { Scrim } from '../scrim/Scrim.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useFloatingUI, useFloatingComponent } from '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import { useId } from '../utils/useId.js';
import { useCurrentBreakpoint } from '../utils/useResponsiveProp.js';
import css_248z from './Dialog.css.js';
import { DialogContext } from './DialogContext.js';

const ConditionalScrimWrapper = ({
  condition,
  children
}) => {
  return condition ? /* @__PURE__ */ jsx(Scrim, { fixed: true, children }) : /* @__PURE__ */ jsxs(Fragment, { children: [
    children,
    " "
  ] });
};
const withBaseName = makePrefixer("saltDialog");
const Dialog = forwardRef(
  function Dialog2(props, ref) {
    var _a, _b;
    const {
      children,
      className,
      open = false,
      onOpenChange,
      status,
      disableDismiss,
      size = "medium",
      disableScrim,
      idProp,
      initialFocus,
      ...rest
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-dialog",
      css: css_248z,
      window: targetWindow
    });
    const id = useId(idProp);
    const currentBreakpoint = useCurrentBreakpoint();
    const [showComponent, setShowComponent] = useState(false);
    const { context, floating, elements } = useFloatingUI({
      open: showComponent,
      onOpenChange
    });
    const { getFloatingProps } = useInteractions([
      useClick(context),
      useDismiss(context, { enabled: !disableDismiss })
    ]);
    const { Component: FloatingComponent } = useFloatingComponent();
    const floatingRef = useForkRef(floating, ref);
    useEffect(() => {
      if (open && !showComponent) {
        setShowComponent(true);
      }
      if (!open && showComponent) {
        const animate = setTimeout(() => {
          setShowComponent(false);
        }, 300);
        return () => clearTimeout(animate);
      }
    }, [open, showComponent, setShowComponent]);
    const contextValue = useMemo(() => ({ status, id }), [status, id]);
    return /* @__PURE__ */ jsx(DialogContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsx(ConditionalScrimWrapper, { condition: showComponent && !disableScrim, children: /* @__PURE__ */ jsx(
      FloatingComponent,
      {
        open: showComponent,
        role: "dialog",
        "aria-modal": "true",
        "aria-labelledby": id,
        ref: floatingRef,
        width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
        height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight,
        lockScroll: true,
        focusManagerProps: {
          context,
          initialFocus
        },
        className: clsx(
          withBaseName(),
          withBaseName(size, currentBreakpoint),
          {
            [withBaseName("enterAnimation")]: open,
            [withBaseName("exitAnimation")]: !open,
            [withBaseName(status)]: status
          },
          className
        ),
        onAnimationEnd: () => {
          if (!open && showComponent) {
            setShowComponent(false);
          }
        },
        ...getFloatingProps(),
        ...rest,
        children
      }
    ) }) });
  }
);

export { ConditionalScrimWrapper, Dialog };
