// @salt-ds/core/dist-cjs/status-adornment/StatusAdornment.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var ErrorAdornment = require('./ErrorAdornment.js');
var SuccessAdornment = require('./SuccessAdornment.js');
var WarningAdornment = require('./WarningAdornment.js');
var StatusAdornment$1 = require('./StatusAdornment.css.js');

const icons = {
  error: ErrorAdornment.ErrorAdornmentIcon,
  warning: WarningAdornment.WarningAdornmentIcon,
  success: SuccessAdornment.SuccessAdornmentIcon
};
const statusToAriaLabelMap = {
  error: "error",
  warning: "warning",
  success: "success"
};
const withBaseName = makePrefixer.makePrefixer("saltStatusAdornment");
const StatusAdornment = React.forwardRef(
  function StatusAdornment2({ className, status, ...restProps }, ref) {
    const AdornmentComponent = icons[status];
    const ariaLabel = statusToAriaLabelMap[status];
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-status-adornment",
      css: StatusAdornment$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(
      AdornmentComponent,
      {
        className: clsx.clsx(withBaseName(), withBaseName(status), className),
        "aria-label": ariaLabel,
        ...restProps,
        ref
      }
    );
  }
);

exports.StatusAdornment = StatusAdornment;
