// @salt-ds/core/dist-es/toast/Toast.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { StatusIndicator } from '../status-indicator/StatusIndicator.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Toast.css.js';

const withBaseName = makePrefixer("saltToast");
const Toast = forwardRef(
  function Toast2(props, ref) {
    const { children, className, status, icon, ...rest } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-toast",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxs(
      "div",
      {
        className: clsx(
          withBaseName(),
          { [withBaseName(status ?? "")]: status },
          className
        ),
        role: "alert",
        ...rest,
        ref,
        children: [
          status && /* @__PURE__ */ jsx("div", { className: withBaseName("iconContainer"), children: icon ? icon : /* @__PURE__ */ jsx(StatusIndicator, { status }) }),
          children
        ]
      }
    );
  }
);

export { Toast };
