// @salt-ds/core/dist-es/tag/Tag.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import { clsx } from 'clsx';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Tag.css.js';

const withBaseName = makePrefixer("saltTag");
const Tag = forwardRef(function Tag2({ children, className, variant = "primary", category = 1, bordered, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tag",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: clsx(
        withBaseName(),
        withBaseName(`category-${category}`),
        withBaseName(variant),
        {
          [withBaseName("bordered")]: bordered
        },
        className
      ),
      ...rest,
      children
    }
  );
});

export { Tag };
