// @salt-ds/core/dist-cjs/option/OptionListBase.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var OptionList = require('./OptionList.css.js');

const withBaseName = makePrefixer.makePrefixer("saltOptionList");
const OptionListBase = React.forwardRef(
  function OptionListBase2(props, ref) {
    const { children, className, ...rest } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-option-list",
      css: OptionList,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: clsx.clsx(withBaseName("container"), className),
        ...rest,
        ref,
        children
      }
    );
  }
);

exports.OptionListBase = OptionListBase;
