// @salt-ds/core/dist-cjs/tooltip/TooltipBase.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('@floating-ui/react');
require('../form-field-context/FormFieldContext.js');
var useFormFieldProps = require('../form-field-context/useFormFieldProps.js');
var StatusIndicator = require('../status-indicator/StatusIndicator.js');
require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('clsx');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Tooltip = require('./Tooltip.css.js');

const withBaseName = makePrefixer.makePrefixer("saltTooltip");
const TooltipBase = (props) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-tooltip",
    css: Tooltip,
    window: targetWindow
  });
  const { a11yProps } = useFormFieldProps.useFormFieldProps();
  const { arrowProps, content, hideArrow, hideIcon, status } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsxs("div", { className: withBaseName("container"), children: [
      !hideIcon && status && /* @__PURE__ */ jsxRuntime.jsx(
        StatusIndicator.StatusIndicator,
        {
          status,
          size: 1,
          className: withBaseName("icon")
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        "span",
        {
          id: a11yProps == null ? void 0 : a11yProps["aria-describedby"],
          className: withBaseName("content"),
          children: content
        }
      )
    ] }),
    !hideArrow && /* @__PURE__ */ jsxRuntime.jsx(
      react.FloatingArrow,
      {
        ...arrowProps,
        className: withBaseName("arrow"),
        strokeWidth: 1,
        fill: "var(--salt-container-primary-background)",
        stroke: "var(--tooltip-status-borderColor)",
        height: 5,
        width: 10
      }
    )
  ] });
};

exports.TooltipBase = TooltipBase;
