// @salt-ds/core/dist-cjs/tooltip/useAriaAnnounce.js
'use strict';

var React = require('react');
require('../aria-announcer/AriaAnnouncerContext.js');
require('react/jsx-runtime');
var useAriaAnnouncer = require('../aria-announcer/useAriaAnnouncer.js');
require('clsx');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
var useIsomorphicLayoutEffect = require('../utils/useIsomorphicLayoutEffect.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');

function getDocument(floating) {
  return (floating == null ? void 0 : floating.ownerDocument) ?? document;
}
function getWindow(value) {
  return getDocument(value).defaultView ?? window;
}
function isElement(value) {
  return value ? value instanceof getWindow(value).Element : false;
}
function getDelay(value, prop, pointerType) {
  if (pointerType && pointerType !== "mouse") {
    return 0;
  }
  if (typeof value === "number") {
    return value;
  }
  return value == null ? void 0 : value[prop];
}
const useAriaAnnounce = (context, { delay = 0 }) => {
  const { open, dataRef, refs } = context;
  const pointerTypeRef = React.useRef();
  const timeoutRef = React.useRef();
  const blockMouseMoveRef = React.useRef(true);
  const { announce } = useAriaAnnouncer.useAriaAnnouncer();
  useIsomorphicLayoutEffect.useIsomorphicLayoutEffect(() => {
    if (!open) {
      pointerTypeRef.current = void 0;
    }
  });
  React.useEffect(() => {
    const reference = refs.reference.current;
    function announceFloating() {
      var _a;
      const tooltipContent = (_a = refs.floating.current) == null ? void 0 : _a.innerText;
      if (tooltipContent) {
        announce(tooltipContent);
      }
    }
    function onMouseEnter(event) {
      clearTimeout(timeoutRef.current);
      if (open) {
        return;
      }
      blockMouseMoveRef.current = false;
      dataRef.current.openEvent = event;
      if (delay) {
        timeoutRef.current = window.setTimeout(
          () => {
            announceFloating();
          },
          getDelay(delay, "open", pointerTypeRef.current)
        );
      } else {
        announceFloating();
      }
    }
    if (isElement(reference)) {
      reference.addEventListener("mouseenter", onMouseEnter);
      return () => {
        reference.removeEventListener("mouseenter", onMouseEnter);
      };
    }
  }, [dataRef, delay, open, refs.reference, refs.floating, announce]);
  function setPointerRef(event) {
    pointerTypeRef.current = event.pointerType;
  }
  return {
    reference: {
      onPointerDown: setPointerRef,
      onPointerEnter: setPointerRef
    },
    floating: {
      onMouseEnter() {
        clearTimeout(timeoutRef.current);
      }
    }
  };
};

exports.useAriaAnnounce = useAriaAnnounce;
