// @salt-ds/core/dist-cjs/tooltip/useTooltip.js
'use strict';

var react = require('@floating-ui/react');
var React = require('react');
require('react/jsx-runtime');
require('clsx');
var useControlled = require('../utils/useControlled.js');
var useFloatingUI = require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var useAriaAnnounce = require('./useAriaAnnounce.js');

function useTooltip(props) {
  const {
    enterDelay,
    disabled,
    leaveDelay,
    open: openProp,
    onOpenChange,
    placement: placementProp,
    disableHoverListener,
    disableFocusListener
  } = props ?? {};
  const arrowRef = React.useRef(null);
  const [open, setOpen] = useControlled.useControlled({
    controlled: openProp,
    default: false,
    name: "Tooltip",
    state: "open"
  });
  const handleOpenChange = (open2) => {
    setOpen(open2);
    onOpenChange == null ? void 0 : onOpenChange(open2);
  };
  const { floating, reference, x, y, strategy, placement, context, elements } = useFloatingUI.useFloatingUI({
    open: disabled ? false : open,
    onOpenChange: handleOpenChange,
    placement: placementProp,
    middleware: [
      react.offset(8),
      react.shift({ limiter: react.limitShift() }),
      react.flip({
        fallbackAxisSideDirection: "end",
        fallbackStrategy: "initialPlacement"
      }),
      react.arrow({ element: arrowRef })
    ]
  });
  const { getReferenceProps, getFloatingProps } = react.useInteractions([
    react.useHover(context, {
      delay: {
        open: enterDelay,
        close: leaveDelay
      },
      enabled: !(disableHoverListener || disabled),
      handleClose: react.safePolygon()
    }),
    react.useFocus(context, { enabled: !(disableFocusListener || disabled) }),
    react.useRole(context, { role: "tooltip" }),
    react.useDismiss(context, { enabled: !disabled }),
    useAriaAnnounce.useAriaAnnounce(context, {
      delay: {
        open: enterDelay,
        close: leaveDelay
      }
    })
  ]);
  const arrowProps = {
    ref: arrowRef,
    context
  };
  const getTooltipProps = () => {
    const { tabIndex, ...tooltipProps } = getFloatingProps({
      // @ts-expect-error - `data-*` props need extra typing when not used on a DOM element.
      "data-placement": placement,
      ref: floating
    });
    return tooltipProps;
  };
  const getTriggerProps = () => getReferenceProps({
    ref: reference
  });
  const getTooltipPosition = () => {
    var _a, _b;
    return {
      top: y ?? 0,
      left: x ?? 0,
      position: strategy,
      width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
      height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight
    };
  };
  return {
    arrowProps,
    open,
    floating,
    reference,
    getTooltipProps,
    getTriggerProps,
    getTooltipPosition
  };
}

exports.useTooltip = useTooltip;
