// @salt-ds/core/dist-cjs/button/Button.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Spinner = require('../spinner/Spinner.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Button$1 = require('./Button.css.js');
var useButton = require('./useButton.js');

const withBaseName = makePrefixer.makePrefixer("saltButton");
const ButtonVariantValues = ["primary", "secondary", "cta"];
const ButtonAppearanceValues = [
  "solid",
  "bordered",
  "transparent"
];
const ButtonSentimentValues = [
  "accented",
  "neutral",
  "positive",
  "negative",
  "caution"
];
function variantToAppearanceAndColor(variant) {
  switch (variant) {
    case "primary":
      return { appearance: "solid", sentiment: "neutral" };
    case "secondary":
      return { appearance: "transparent", sentiment: "neutral" };
    case "cta":
      return { appearance: "solid", sentiment: "accented" };
  }
}
const Button = React.forwardRef(
  function Button2({
    children,
    className,
    disabled,
    focusableWhenDisabled,
    onKeyUp,
    onKeyDown,
    onBlur,
    onClick,
    loading,
    loadingAnnouncement,
    appearance: appearanceProp,
    sentiment: sentimentProp,
    type = "button",
    variant = "primary",
    ...restProps
  }, ref) {
    const { active, buttonProps } = useButton.useButton({
      loading,
      disabled,
      focusableWhenDisabled,
      onKeyUp,
      onKeyDown,
      onBlur,
      onClick
    });
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-button",
      css: Button$1,
      window: targetWindow
    });
    const mapped = variantToAppearanceAndColor(variant);
    const appearance = appearanceProp ?? (mapped == null ? void 0 : mapped.appearance) ?? "solid";
    const sentiment = sentimentProp ?? (mapped == null ? void 0 : mapped.sentiment) ?? "neutral";
    const { tabIndex, ...restButtonProps } = buttonProps;
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "button",
      {
        ...restButtonProps,
        className: clsx.clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("loading")]: loading,
            [withBaseName("disabled")]: disabled,
            [withBaseName("active")]: active,
            [withBaseName(appearance)]: appearance,
            [withBaseName(sentiment)]: sentiment
          },
          className
        ),
        ...restProps,
        ref,
        type,
        children: [
          loading && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("spinner"), "aria-hidden": true, children: /* @__PURE__ */ jsxRuntime.jsx(Spinner.Spinner, { size: "small", "aria-hidden": true, disableAnnouncer: true }) }),
          typeof loadingAnnouncement === "string" && /* @__PURE__ */ jsxRuntime.jsx("span", { role: "status", className: withBaseName("sr-only"), children: loadingAnnouncement }),
          children
        ]
      }
    );
  }
);

exports.Button = Button;
exports.ButtonAppearanceValues = ButtonAppearanceValues;
exports.ButtonSentimentValues = ButtonSentimentValues;
exports.ButtonVariantValues = ButtonVariantValues;
