// @salt-ds/core/dist-cjs/dialog/DialogCloseButton.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Button = require('../button/Button.js');
var SemanticIconProvider = require('../semantic-icon-provider/SemanticIconProvider.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var DialogCloseButton$1 = require('./DialogCloseButton.css.js');

const withBaseName = makePrefixer.makePrefixer("saltDialogCloseButton");
const DialogCloseButton = React.forwardRef(
  function DialogCloseButton2({ className, ...rest }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-dialog-close-button",
      css: DialogCloseButton$1,
      window: targetWindow
    });
    const { CloseIcon } = SemanticIconProvider.useIcon();
    return /* @__PURE__ */ jsxRuntime.jsx(
      Button.Button,
      {
        ref,
        "aria-label": "Close dialog",
        appearance: "transparent",
        className: clsx(withBaseName(), className),
        ...rest,
        children: /* @__PURE__ */ jsxRuntime.jsx(CloseIcon, { "aria-hidden": true })
      }
    );
  }
);

exports.DialogCloseButton = DialogCloseButton;
