// @salt-ds/core/dist-cjs/dialog/Dialog.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('@floating-ui/react');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Scrim = require('../scrim/Scrim.js');
var makePrefixer = require('../utils/makePrefixer.js');
var useFloatingUI = require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
var useId = require('../utils/useId.js');
var useResponsiveProp = require('../utils/useResponsiveProp.js');
var Dialog$1 = require('./Dialog.css.js');
var DialogContext = require('./DialogContext.js');

const ConditionalScrimWrapper = ({
  condition,
  children
}) => {
  return condition ? /* @__PURE__ */ jsxRuntime.jsx(Scrim.Scrim, { fixed: true, children }) : /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    children,
    " "
  ] });
};
const withBaseName = makePrefixer.makePrefixer("saltDialog");
const Dialog = React.forwardRef(
  function Dialog2(props, ref) {
    var _a, _b;
    const {
      children,
      className,
      open = false,
      onOpenChange,
      status,
      disableDismiss,
      size = "medium",
      disableScrim,
      idProp,
      initialFocus,
      ...rest
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-dialog",
      css: Dialog$1,
      window: targetWindow
    });
    const id = useId.useId(idProp);
    const currentBreakpoint = useResponsiveProp.useCurrentBreakpoint();
    const [showComponent, setShowComponent] = React.useState(false);
    const { context, floating, elements } = useFloatingUI.useFloatingUI({
      open: showComponent,
      onOpenChange
    });
    const { getFloatingProps } = react.useInteractions([
      react.useClick(context),
      react.useDismiss(context, { enabled: !disableDismiss })
    ]);
    const { Component: FloatingComponent } = useFloatingUI.useFloatingComponent();
    const floatingRef = useForkRef.useForkRef(floating, ref);
    React.useEffect(() => {
      if (open && !showComponent) {
        setShowComponent(true);
      }
      if (!open && showComponent) {
        const animate = setTimeout(() => {
          setShowComponent(false);
        }, 300);
        return () => clearTimeout(animate);
      }
    }, [open, showComponent, setShowComponent]);
    const contextValue = React.useMemo(() => ({ status, id }), [status, id]);
    return /* @__PURE__ */ jsxRuntime.jsx(DialogContext.DialogContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsxRuntime.jsx(ConditionalScrimWrapper, { condition: showComponent && !disableScrim, children: /* @__PURE__ */ jsxRuntime.jsx(
      FloatingComponent,
      {
        open: showComponent,
        role: "dialog",
        "aria-modal": "true",
        "aria-labelledby": id,
        ref: floatingRef,
        width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
        height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight,
        lockScroll: true,
        focusManagerProps: {
          context,
          initialFocus
        },
        className: clsx.clsx(
          withBaseName(),
          withBaseName(size, currentBreakpoint),
          {
            [withBaseName("enterAnimation")]: open,
            [withBaseName("exitAnimation")]: !open,
            [withBaseName(status)]: status
          },
          className
        ),
        onAnimationEnd: () => {
          if (!open && showComponent) {
            setShowComponent(false);
          }
        },
        ...getFloatingProps(),
        ...rest,
        children
      }
    ) }) });
  }
);

exports.ConditionalScrimWrapper = ConditionalScrimWrapper;
exports.Dialog = Dialog;
