// @salt-ds/core/dist-cjs/banner/Banner.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var StatusIndicator = require('../status-indicator/StatusIndicator.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var Banner$1 = require('./Banner.css.js');

const withBaseName = makePrefixer.makePrefixer("saltBanner");
const Banner = React.forwardRef(function Banner2({ children, className, variant = "primary", status = "info", ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-banner",
    css: Banner$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      className: clsx.clsx(
        withBaseName(),
        withBaseName(status),
        withBaseName(variant),
        className
      ),
      ref,
      ...rest,
      "aria-live": "polite",
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(StatusIndicator.StatusIndicator, { status, className: withBaseName("icon") }),
        children
      ]
    }
  );
});

exports.Banner = Banner;
