// @salt-ds/core/dist-cjs/status-indicator/StatusIndicator.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var icons = require('@salt-ds/icons');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var SemanticIconProvider = require('../semantic-icon-provider/SemanticIconProvider.js');
var StatusIndicator$1 = require('./StatusIndicator.css.js');

const statusToAriaLabelMap = {
  error: "error",
  success: "success",
  warning: "warning",
  info: "info"
};
const withBaseName = makePrefixer.makePrefixer("saltStatusIndicator");
const StatusIndicator = React.forwardRef(
  function StatusIndicator2({ className, status, size = icons.DEFAULT_ICON_SIZE, ...restProps }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-status-indicator",
      css: StatusIndicator$1,
      window: targetWindow
    });
    const { ErrorIcon, WarningIcon, SuccessIcon, InfoIcon } = SemanticIconProvider.useIcon();
    const iconMap = {
      error: ErrorIcon,
      success: SuccessIcon,
      warning: WarningIcon,
      info: InfoIcon
    };
    const IconComponent = iconMap[status];
    const ariaLabel = statusToAriaLabelMap[status];
    return /* @__PURE__ */ jsxRuntime.jsx(
      IconComponent,
      {
        className: clsx.clsx(withBaseName(), withBaseName(status), className),
        size,
        "aria-label": ariaLabel,
        ...restProps,
        ref
      }
    );
  }
);

exports.StatusIndicator = StatusIndicator;
