// @salt-ds/core/dist-cjs/multiline-input/MultilineInput.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
require('../form-field-context/FormFieldContext.js');
var useFormFieldProps = require('../form-field-context/useFormFieldProps.js');
var StatusAdornment = require('../status-adornment/StatusAdornment.js');
var makePrefixer = require('../utils/makePrefixer.js');
var useControlled = require('../utils/useControlled.js');
require('../utils/useFloatingUI/useFloatingUI.js');
var useForkRef = require('../utils/useForkRef.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var MultilineInput$1 = require('./MultilineInput.css.js');

const withBaseName = makePrefixer.makePrefixer("saltMultilineInput");
const MultilineInput = React.forwardRef(
  function MultilineInput2({
    "aria-activedescendant": ariaActiveDescendant,
    "aria-expanded": ariaExpanded,
    "aria-owns": ariaOwns,
    bordered = false,
    className: classNameProp,
    disabled,
    endAdornment,
    id,
    placeholder,
    readOnly,
    role,
    rows = 3,
    startAdornment,
    style,
    textAreaProps = {},
    textAreaRef,
    value: valueProp,
    defaultValue: defaultValueProp = valueProp === void 0 ? "" : void 0,
    validationStatus: validationStatusProp,
    variant = "primary",
    ...other
  }, ref) {
    const inputRef = React.useRef(null);
    const handleRef = useForkRef.useForkRef(inputRef, textAreaRef);
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-multiline-input",
      css: MultilineInput$1,
      window: targetWindow
    });
    const restA11yProps = {
      "aria-activedescendant": ariaActiveDescendant,
      "aria-expanded": ariaExpanded,
      "aria-owns": ariaOwns
    };
    const [focused, setFocused] = React.useState(false);
    const {
      "aria-describedby": textAreaDescribedBy,
      "aria-labelledby": textAreaLabelledBy,
      onBlur,
      onChange,
      onFocus,
      required: textAreaRequired,
      ...restTextAreaProps
    } = textAreaProps;
    const {
      a11yProps: {
        "aria-describedby": formFieldDescribedBy,
        "aria-labelledby": formFieldLabelledBy
      } = {},
      disabled: formFieldDisabled,
      readOnly: formFieldReadOnly,
      necessity: formFieldRequired,
      validationStatus: formFieldValidationStatus
    } = useFormFieldProps.useFormFieldProps();
    const isDisabled = disabled || formFieldDisabled;
    const isReadOnly = readOnly || formFieldReadOnly;
    const validationStatus = formFieldValidationStatus ?? validationStatusProp;
    const isRequired = formFieldRequired ? ["required", "asterisk"].includes(formFieldRequired) : textAreaRequired;
    const [value, setValue] = useControlled.useControlled({
      controlled: valueProp,
      default: defaultValueProp,
      name: "MultilineInput",
      state: "value"
    });
    const previousHeight = React.useRef(void 0);
    const input = inputRef.current;
    const changeHeight = React.useCallback(() => {
      if (!input) return;
      const hasBeenManuallyResized = previousHeight.current !== void 0 && input.style.height !== previousHeight.current;
      if (!hasBeenManuallyResized) {
        const previousOverflow = input.style.overflow;
        input.style.overflow = "hidden";
        input.style.height = "auto";
        const newHeight = `${input.scrollHeight + (input.offsetHeight - input.clientHeight)}px`;
        input.style.height = newHeight;
        previousHeight.current = newHeight;
        input.style.overflow = previousOverflow;
      }
    }, [input]);
    const handleChange = (event) => {
      const value2 = event.target.value;
      setValue(value2);
      onChange == null ? void 0 : onChange(event);
      changeHeight();
    };
    React.useLayoutEffect(() => {
      changeHeight();
    }, [value, changeHeight]);
    const handleBlur = (event) => {
      onBlur == null ? void 0 : onBlur(event);
      setFocused(false);
    };
    const handleFocus = (event) => {
      onFocus == null ? void 0 : onFocus(event);
      setFocused(true);
    };
    const multilineInputStyles = {
      "--saltMultilineInput-rows": rows,
      ...style
    };
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "div",
      {
        className: clsx.clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("withAdornmentRow")]: endAdornment,
            [withBaseName("bordered")]: bordered,
            [withBaseName("focused")]: !isDisabled && focused,
            [withBaseName("disabled")]: isDisabled,
            [withBaseName("readOnly")]: isReadOnly,
            [withBaseName(validationStatus ?? "")]: validationStatus
          },
          classNameProp
        ),
        ref,
        style: multilineInputStyles,
        ...other,
        children: [
          startAdornment && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("startAdornmentContainer"), children: startAdornment }),
          /* @__PURE__ */ jsxRuntime.jsx(
            "textarea",
            {
              "aria-describedby": clsx.clsx(formFieldDescribedBy, textAreaDescribedBy),
              "aria-labelledby": clsx.clsx(formFieldLabelledBy, textAreaLabelledBy),
              className: clsx.clsx(withBaseName("textarea"), textAreaProps == null ? void 0 : textAreaProps.className),
              disabled: isDisabled,
              id,
              readOnly: isReadOnly,
              ref: handleRef,
              required: isRequired,
              role,
              rows,
              tabIndex: isDisabled ? -1 : 0,
              onBlur: handleBlur,
              onChange: handleChange,
              onFocus: !isDisabled ? handleFocus : void 0,
              placeholder,
              value,
              ...restA11yProps,
              ...restTextAreaProps
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: withBaseName("suffixAdornments"), children: [
            !isDisabled && validationStatus && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("statusAdornmentContainer"), children: /* @__PURE__ */ jsxRuntime.jsx(StatusAdornment.StatusAdornment, { status: validationStatus }) }),
            endAdornment && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("endAdornmentContainer"), children: endAdornment })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("activationIndicator") })
        ]
      }
    );
  }
);

exports.MultilineInput = MultilineInput;
