// @salt-ds/core/dist-es/scrim/Scrim.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './Scrim.css.js';

const withBaseName = makePrefixer("saltScrim");
const Scrim = forwardRef(function Scrim2({ className, children, fixed = false, open = true, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-scrim",
    css: css_248z,
    window: targetWindow
  });
  if (!open) {
    return null;
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: clsx(
        withBaseName(),
        {
          [withBaseName("fixed")]: fixed
        },
        className
      ),
      "data-testid": "scrim",
      ref,
      ...rest,
      children
    }
  );
});

export { Scrim };
