// @salt-ds/core/dist-cjs/link-card/LinkCard.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var capitalize = require('../utils/capitalize.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var LinkCard$1 = require('./LinkCard.css.js');

const withBaseName = makePrefixer.makePrefixer("saltLinkCard");
const LinkCard = React.forwardRef(
  function LinkCard2(props, ref) {
    const {
      accent,
      children,
      className,
      href,
      variant = "primary",
      ...rest
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-link-card",
      css: LinkCard$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(
      "a",
      {
        className: clsx.clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("accent")]: accent,
            [withBaseName(`accent${capitalize.capitalize(accent ?? "")}`)]: accent
          },
          className
        ),
        href,
        ...rest,
        ref,
        children
      }
    );
  }
);

exports.LinkCard = LinkCard;
