// @salt-ds/core/dist-es/pagination/CompactInput.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef, useState, useEffect } from 'react';
import { Input } from '../input/Input.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { usePaginationContext } from './usePaginationContext.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './CompactInput.css.js';

const withBaseName = makePrefixer("saltCompactInput");
const CompactInput = forwardRef(function CompactInput2(props, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-compact-input",
    css: css_248z,
    window: targetWindow
  });
  const { count, page, onPageChange } = usePaginationContext();
  const [inputValue, setInputValue] = useState(`${page}`);
  useEffect(() => {
    setInputValue(`${page}`);
  }, [page]);
  const handleChange = (event) => {
    setInputValue(event.target.value);
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      const pageValue = Number(inputValue);
      if (!Number.isNaN(pageValue) && pageValue <= count && pageValue > 0) {
        onPageChange(event, pageValue);
      } else {
        setInputValue(`${page}`);
      }
    }
  };
  const handleBlur = () => {
    setInputValue(`${page}`);
  };
  return /* @__PURE__ */ jsx(
    Input,
    {
      className: clsx(withBaseName(), {
        [withBaseName("defaultSize")]: count < 100
      }),
      inputProps: {
        "aria-label": `Go to page, ${count} total`,
        style: { width: `${`${count}`.length}ch` }
      },
      onBlur: handleBlur,
      onChange: handleChange,
      onKeyDown: handleKeyDown,
      value: inputValue,
      textAlign: "center",
      ref,
      ...props
    }
  );
});

export { CompactInput };
