// @salt-ds/core/dist-es/pagination/Paginator.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { Button } from '../button/Button.js';
import { useIcon } from '../semantic-icon-provider/SemanticIconProvider.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { PageRanges } from './PageRanges.js';
import { usePaginationContext } from './usePaginationContext.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './Paginator.css.js';

const withBaseName = makePrefixer("saltPaginator");
const Paginator = forwardRef(
  function Paginator2({ className, boundaryCount, siblingCount, ...restProps }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-paginator",
      css: css_248z,
      window: targetWindow
    });
    const { count, page, onPageChange } = usePaginationContext();
    const { NextIcon, PreviousIcon } = useIcon();
    const onPreviousPage = (event) => {
      onPageChange(event, Math.max(1, page - 1));
    };
    const onNextPage = (event) => {
      onPageChange(event, Math.min(page + 1, count));
    };
    const isOnFirstPage = page === 1;
    const isOnLastPage = page === count;
    return /* @__PURE__ */ jsxs("div", { className: clsx(withBaseName(), className), ...restProps, ref, children: [
      /* @__PURE__ */ jsx(
        Button,
        {
          appearance: "transparent",
          "aria-label": "Previous Page",
          onClick: onPreviousPage,
          disabled: isOnFirstPage,
          className: withBaseName("arrowButton-previous"),
          children: /* @__PURE__ */ jsx(PreviousIcon, { "aria-hidden": true })
        }
      ),
      /* @__PURE__ */ jsx(PageRanges, { siblingCount, boundaryCount }),
      /* @__PURE__ */ jsx(
        Button,
        {
          appearance: "transparent",
          "aria-label": "Next Page",
          onClick: onNextPage,
          disabled: isOnLastPage,
          className: withBaseName("arrowButton-next"),
          children: /* @__PURE__ */ jsx(NextIcon, { "aria-hidden": true })
        }
      )
    ] });
  }
);

export { Paginator };
