// @salt-ds/core/dist-es/stepper/Stepper.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useContext } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Stepper.css.js';
import { StepperOrientationContext, StepperProvider } from './internal/StepperProvider.js';

const withBaseName = makePrefixer("saltStepper");
const Stepper = forwardRef(
  function Stepper2({ orientation: orientationProp, children, className, ...props }, ref) {
    const targetWindow = useWindow();
    const orientationContext = useContext(StepperOrientationContext);
    const orientation = orientationProp || orientationContext;
    useComponentCssInjection({
      testId: "salt-Stepper",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(StepperProvider, { orientation, children: /* @__PURE__ */ jsx(
      "ol",
      {
        className: clsx(withBaseName(), withBaseName(orientation), className),
        ref,
        ...props,
        children
      }
    ) });
  }
);

export { Stepper };
