// @salt-ds/core/dist-cjs/border-item/BorderItem.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var GridItem = require('../grid-item/GridItem.js');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var BorderItem$1 = require('./BorderItem.css.js');

const BORDER_POSITION = [
  "north",
  "west",
  "center",
  "east",
  "south"
];
const withBaseName = makePrefixer.makePrefixer("saltBorderItem");
const BorderItem = React.forwardRef(
  ({
    children,
    className,
    position,
    sticky = false,
    style,
    ...rest
  }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-border-item",
      css: BorderItem$1,
      window: targetWindow
    });
    const gridItemStyles = {
      ...style,
      "--gridItem-gridArea": position
    };
    return /* @__PURE__ */ jsxRuntime.jsx(
      GridItem.GridItem,
      {
        ref,
        className: clsx.clsx(
          withBaseName(),
          "saltGridItem-area",
          {
            [withBaseName("sticky")]: sticky
          },
          className
        ),
        style: gridItemStyles,
        ...rest,
        children
      }
    );
  }
);

exports.BORDER_POSITION = BORDER_POSITION;
exports.BorderItem = BorderItem;
