// @salt-ds/core/dist-cjs/divider/Divider.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Divider$1 = require('./Divider.css.js');

const withBaseName = makePrefixer.makePrefixer("saltDivider");
const Divider = React.forwardRef(
  function Divider2(props, ref) {
    const {
      className,
      orientation = "horizontal",
      variant = "primary",
      ...rest
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-divider",
      css: Divider$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: clsx.clsx(
          withBaseName(),
          withBaseName(orientation),
          withBaseName(variant),
          className
        ),
        "aria-orientation": orientation,
        role: "separator",
        ref,
        ...rest
      }
    );
  }
);

exports.Divider = Divider;
