// @salt-ds/core/dist-es/link-card/LinkCard.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { capitalize } from '../utils/capitalize.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './LinkCard.css.js';

const withBaseName = makePrefixer("saltLinkCard");
const LinkCard = forwardRef(
  function LinkCard2(props, ref) {
    const {
      accent,
      children,
      className,
      href,
      variant = "primary",
      ...rest
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-link-card",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(
      "a",
      {
        className: clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("accent")]: accent,
            [withBaseName(`accent${capitalize(accent ?? "")}`)]: accent
          },
          className
        ),
        href,
        ...rest,
        ref,
        children
      }
    );
  }
);

export { LinkCard };
