// @salt-ds/core/dist-es/list-control/ListControlState.js
import { useState, useEffect, useRef, useCallback } from 'react';
import 'react/jsx-runtime';
import 'clsx';
import { useControlled } from '../utils/useControlled.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';

function findElementPosition(elements, element) {
  if (elements.length === 0) {
    return [0, false];
  }
  if (element.compareDocumentPosition(elements[elements.length - 1].element) & Node.DOCUMENT_POSITION_PRECEDING) {
    return [-1, false];
  }
  if (element.compareDocumentPosition(elements[0].element) & Node.DOCUMENT_POSITION_FOLLOWING) {
    return [0, false];
  }
  let left = 0;
  let right = elements.length;
  let leftLast = 0;
  let rightLast = right;
  let exists = false;
  while (left < right) {
    const inPos = Math.floor((right + left) / 2);
    const compared = element.compareDocumentPosition(elements[inPos].element);
    if (compared & Node.DOCUMENT_POSITION_PRECEDING) {
      left = inPos;
    } else if (compared & Node.DOCUMENT_POSITION_FOLLOWING) {
      right = inPos;
    } else {
      right = inPos;
      left = inPos;
      exists = true;
    }
    if (leftLast === left && rightLast === right) {
      break;
    }
    leftLast = left;
    rightLast = right;
  }
  return [right, exists];
}
function defaultValueToString(item) {
  return String(item);
}
function useListControl(props) {
  const {
    open: openProp,
    defaultOpen,
    onOpenChange,
    multiselect,
    defaultSelected,
    selected: selectedProp,
    onSelectionChange,
    disabled,
    readOnly,
    valueToString = defaultValueToString
  } = props;
  const [focusedState, setFocusedState] = useState(false);
  const [focusVisibleState, setFocusVisibleState] = useState(false);
  useEffect(() => {
    if (disabled && focusedState) {
      setFocusedState(false);
      setFocusVisibleState(false);
    }
  }, [disabled, focusedState]);
  const [activeState, setActiveState] = useState(
    void 0
  );
  const setActive = (option) => {
    if (option) {
      setActiveState(option);
    } else {
      setActiveState(void 0);
    }
  };
  const [openState, setOpenState] = useControlled({
    controlled: openProp,
    default: Boolean(defaultOpen),
    name: "ListControl",
    state: "open"
  });
  const openKey = useRef(void 0);
  const setOpen = (newOpen, reason, key) => {
    if (disabled || readOnly) {
      return;
    }
    setOpenState(newOpen);
    openKey.current = key;
    if (newOpen !== openState) {
      onOpenChange == null ? void 0 : onOpenChange(newOpen, reason);
    }
  };
  const [selectedState, setSelectedState] = useControlled({
    controlled: selectedProp,
    default: defaultSelected ?? [],
    name: "ListControl",
    state: "selected"
  });
  const select = (event, option) => {
    if (option.disabled || readOnly || disabled) {
      return;
    }
    let newSelected = [option.value];
    if (multiselect) {
      if (selectedState.includes(option.value)) {
        newSelected = selectedState.filter((item) => item !== option.value);
      } else {
        newSelected = selectedState.concat([option.value]);
      }
    }
    setSelectedState(newSelected);
    onSelectionChange == null ? void 0 : onSelectionChange(event, newSelected);
    if (!multiselect) {
      setOpen(false);
    }
  };
  const clear = (event) => {
    setSelectedState([]);
    if (selectedState.length !== 0) {
      onSelectionChange == null ? void 0 : onSelectionChange(event, []);
    }
  };
  const optionsRef = useRef([]);
  const register = useCallback(
    (optionValue, element) => {
      const { id } = optionValue;
      const [index, exists] = findElementPosition(optionsRef.current, element);
      if (!exists) {
        if (index === -1) {
          optionsRef.current.push({ data: optionValue, element });
        } else {
          optionsRef.current.splice(index, 0, { data: optionValue, element });
        }
      }
      return () => {
        optionsRef.current = optionsRef.current.filter(
          (item) => item.data.id !== id
        );
      };
    },
    []
  );
  const getOptionAtIndex = (index) => {
    return optionsRef.current[index];
  };
  const getIndexOfOption = (option) => {
    return optionsRef.current.findIndex((item) => item.data.id === option.id);
  };
  const getOptionsMatching = (predicate) => {
    return optionsRef.current.filter((item) => predicate(item.data));
  };
  const getOptionFromSearch = (search, startFrom) => {
    const collator = new Intl.Collator("en", {
      usage: "search",
      sensitivity: "base"
    });
    const startIndex = startFrom ? getIndexOfOption(startFrom) + 1 : 0;
    const searchList = optionsRef.current.map((item) => item.data);
    let matches = searchList.filter(
      (option) => collator.compare(
        valueToString(option.value).substring(0, search.length),
        search
      ) === 0
    );
    if (matches.length === 0) {
      const letters = search.split("");
      const allSameLetter = letters.length > 0 && letters.every((letter) => collator.compare(letter, letters[0]) === 0);
      if (allSameLetter) {
        matches = searchList.filter(
          (option) => collator.compare(
            valueToString(option.value)[0].toLowerCase(),
            letters[0]
          ) === 0
        );
      }
    }
    return matches.find((option) => getIndexOfOption(option) >= startIndex);
  };
  const getFirstOption = () => {
    return getOptionAtIndex(0);
  };
  const getLastOption = () => {
    return getOptionAtIndex(optionsRef.current.length - 1);
  };
  const getOptionBefore = (option) => {
    const index = getIndexOfOption(option);
    return getOptionAtIndex(index - 1);
  };
  const getOptionAfter = (option) => {
    const index = getIndexOfOption(option);
    return getOptionAtIndex(index + 1);
  };
  const listRef = useRef(null);
  const getOptionPageAbove = (start) => {
    var _a;
    const list = listRef.current;
    let option = optionsRef.current.find((option2) => option2.data === start);
    if (!list || !option) {
      return void 0;
    }
    const containerRect = list.getBoundingClientRect();
    let optionRect = option.element.getBoundingClientRect();
    const listY = containerRect.y - list.scrollTop;
    const pageY = Math.max(
      0,
      optionRect.y - listY + optionRect.height - containerRect.height
    );
    while (option && optionRect && optionRect.y - listY > pageY) {
      option = getOptionBefore(option.data);
      optionRect = (_a = option == null ? void 0 : option.element) == null ? void 0 : _a.getBoundingClientRect();
    }
    return option ?? getFirstOption();
  };
  const getOptionPageBelow = (start) => {
    const list = listRef.current;
    let option = optionsRef.current.find((option2) => option2.data === start);
    if (!list || !option) {
      return void 0;
    }
    const containerRect = list.getBoundingClientRect();
    let optionRect = option.element.getBoundingClientRect();
    const listY = containerRect.y - list.scrollTop;
    const pageY = Math.min(
      list.scrollHeight,
      optionRect.y - listY - optionRect.height + containerRect.height
    );
    while (option && optionRect && optionRect.y - listY < pageY) {
      option = getOptionAfter(option.data);
      optionRect = option == null ? void 0 : option.element.getBoundingClientRect();
    }
    return option ?? getLastOption();
  };
  useEffect(() => {
    var _a;
    if (listRef.current) {
      const activeElement = (_a = optionsRef.current.find(
        (option) => option.data === activeState
      )) == null ? void 0 : _a.element;
      if (!activeElement) {
        return;
      }
      activeElement.scrollIntoView({
        block: "nearest",
        inline: "nearest"
      });
    }
  }, [activeState]);
  return {
    multiselect: Boolean(multiselect),
    openState,
    setOpen,
    openKey,
    activeState,
    setActive,
    selectedState,
    setSelectedState,
    select,
    clear,
    focusVisibleState,
    setFocusVisibleState,
    focusedState,
    setFocusedState,
    listRef,
    options: optionsRef.current.map((option) => option.element),
    register,
    getOptionAtIndex,
    getIndexOfOption,
    getOptionsMatching,
    getOptionFromSearch,
    getOptionAfter,
    getOptionBefore,
    getOptionPageAbove,
    getOptionPageBelow,
    getFirstOption,
    getLastOption,
    valueToString,
    disabled
  };
}

export { defaultValueToString, useListControl };
