// @salt-ds/core/dist-es/stepper/internal/StepText.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { Text } from '../../text/Text.js';
import '../../text/Code.js';
import '../../text/Display.js';
import '../../text/Headings.js';
import '../../text/Label.js';
import '../../text/TextAction.js';
import '../../text/TextNotation.js';
import 'react';
import { makePrefixer } from '../../utils/makePrefixer.js';
import '../../utils/useFloatingUI/useFloatingUI.js';
import '../../utils/useId.js';
import '../../salt-provider/SaltProvider.js';
import '../../viewport/ViewportProvider.js';
import css_248z from './StepText.css.js';

const withBaseName = makePrefixer("saltStepText");
function StepText({
  id,
  purpose,
  className,
  styleAs = "label",
  children,
  ...props
}) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-step-text",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx(
    Text,
    {
      id,
      styleAs: "label",
      className: clsx(withBaseName(), withBaseName(purpose), className),
      ...props,
      children
    }
  );
}

export { StepText };
