// @salt-ds/core/dist-es/stepper/internal/StepExpandTrigger.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { Button } from '../../button/Button.js';
import 'react';
import { useIcon } from '../../semantic-icon-provider/SemanticIconProvider.js';
import { makePrefixer } from '../../utils/makePrefixer.js';
import '../../utils/useFloatingUI/useFloatingUI.js';
import '../../utils/useId.js';
import '../../salt-provider/SaltProvider.js';
import '../../viewport/ViewportProvider.js';
import css_248z from './StepExpandTrigger.css.js';

const withBaseName = makePrefixer("saltStepExpandTrigger");
function StepExpandTrigger({
  id,
  expanded,
  className,
  ...props
}) {
  const { CollapseIcon, ExpandIcon } = useIcon();
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-step-expand-trigger",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx(
    Button,
    {
      id,
      sentiment: "neutral",
      appearance: "transparent",
      className: clsx(withBaseName(), className),
      ...props,
      children: expanded ? /* @__PURE__ */ jsx(CollapseIcon, { "aria-hidden": true }) : /* @__PURE__ */ jsx(ExpandIcon, { "aria-hidden": true })
    }
  );
}

export { StepExpandTrigger };
