// @salt-ds/core/dist-es/checkbox/Checkbox.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef, useRef } from 'react';
import '../form-field-context/FormFieldContext.js';
import { useFormFieldProps } from '../form-field-context/useFormFieldProps.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useControlled } from '../utils/useControlled.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import '../utils/useId.js';
import { useIsomorphicLayoutEffect } from '../utils/useIsomorphicLayoutEffect.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { CheckboxIcon } from './CheckboxIcon.js';
import { useCheckboxGroup } from './internal/useCheckboxGroup.js';
import css_248z from './Checkbox.css.js';

const withBaseName = makePrefixer("saltCheckbox");
const Checkbox = forwardRef(
  function Checkbox2({
    checked: checkedProp,
    className,
    defaultChecked,
    disabled: disabledProp,
    error,
    indeterminate,
    inputProps = {},
    inputRef: inputRefProp,
    label,
    name,
    onBlur,
    onChange,
    onFocus,
    value,
    validationStatus: validationStatusProp,
    readOnly: readOnlyProp,
    ...rest
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-checkbox",
      css: css_248z,
      window: targetWindow
    });
    const checkboxGroup = useCheckboxGroup();
    const {
      "aria-describedby": inputDescribedBy,
      "aria-labelledby": inputLabelledBy,
      className: inputClassName,
      onChange: inputOnChange,
      ...restInputProps
    } = inputProps;
    const checkboxGroupChecked = checkedProp ?? ((checkboxGroup == null ? void 0 : checkboxGroup.checkedValues) && value ? checkboxGroup.checkedValues.includes(value) : checkedProp);
    const [checked, setChecked] = useControlled({
      controlled: checkboxGroupChecked,
      default: Boolean(defaultChecked),
      name: "Checkbox",
      state: "checked"
    });
    const {
      a11yProps: formFieldA11yProps,
      disabled: formFieldDisabled,
      readOnly: formFieldReadOnly,
      validationStatus: formFieldValidationStatus
    } = useFormFieldProps();
    const disabled = (checkboxGroup == null ? void 0 : checkboxGroup.disabled) || formFieldDisabled || disabledProp;
    const readOnly = (checkboxGroup == null ? void 0 : checkboxGroup.readOnly) || formFieldReadOnly || readOnlyProp;
    const validationStatus = !disabled ? (checkboxGroup == null ? void 0 : checkboxGroup.validationStatus) ?? formFieldValidationStatus ?? validationStatusProp : void 0;
    const inputRef = useRef(null);
    const handleInputRef = useForkRef(inputRefProp, inputRef);
    const handleChange = (event) => {
      var _a;
      if (event.nativeEvent.defaultPrevented || readOnly) {
        return;
      }
      const value2 = event.target.checked;
      setChecked(value2);
      onChange == null ? void 0 : onChange(event);
      inputOnChange == null ? void 0 : inputOnChange(event);
      (_a = checkboxGroup == null ? void 0 : checkboxGroup.onChange) == null ? void 0 : _a.call(checkboxGroup, event);
    };
    useIsomorphicLayoutEffect(() => {
      if (inputRef.current != null) {
        inputRef.current.indeterminate = indeterminate ?? false;
      }
    }, [indeterminate]);
    return /* @__PURE__ */ jsxs(
      "label",
      {
        className: clsx(
          withBaseName(),
          {
            [withBaseName("disabled")]: disabled,
            [withBaseName("readOnly")]: readOnly,
            [withBaseName("error")]: error,
            [withBaseName(validationStatus ?? "")]: validationStatus
          },
          className
        ),
        ref,
        ...rest,
        children: [
          /* @__PURE__ */ jsx(
            "input",
            {
              "aria-readonly": readOnly || void 0,
              "aria-describedby": clsx(
                checkboxGroup === void 0 ? formFieldA11yProps == null ? void 0 : formFieldA11yProps["aria-describedby"] : void 0,
                inputDescribedBy
              ),
              "aria-labelledby": clsx(
                checkboxGroup === void 0 ? formFieldA11yProps == null ? void 0 : formFieldA11yProps["aria-labelledby"] : void 0,
                inputLabelledBy
              ),
              name,
              value,
              checked,
              className: clsx(withBaseName("input"), inputClassName),
              "data-indeterminate": indeterminate,
              defaultChecked,
              disabled,
              readOnly,
              onBlur,
              onChange: handleChange,
              onFocus,
              type: "checkbox",
              ref: handleInputRef,
              ...restInputProps
            }
          ),
          /* @__PURE__ */ jsx(
            CheckboxIcon,
            {
              checked,
              disabled,
              readOnly,
              indeterminate,
              validationStatus,
              error
            }
          ),
          label
        ]
      }
    );
  }
);

export { Checkbox };
