// @salt-ds/core/dist-cjs/progress/CircularProgress/CircularProgress.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var Text = require('../../text/Text.js');
require('../../text/Code.js');
require('../../text/Display.js');
require('../../text/Headings.js');
require('../../text/Label.js');
require('../../text/TextAction.js');
require('../../text/TextNotation.js');
var makePrefixer = require('../../utils/makePrefixer.js');
require('../../utils/useFloatingUI/useFloatingUI.js');
require('../../utils/useId.js');
require('../../salt-provider/SaltProvider.js');
require('../../viewport/ViewportProvider.js');
var CircularProgress$1 = require('./CircularProgress.css.js');

const withBaseName = makePrefixer.makePrefixer("saltCircularProgress");
const getRotationAngle = (bar, shift = 0) => {
  return -180 + (bar - shift) / 50 * 180;
};
const CircularProgress = React.forwardRef(function CircularProgress2({
  className,
  hideLabel = false,
  max = 100,
  min = 0,
  value = 0,
  bufferValue = 0,
  ...rest
}, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-circular-progress",
    css: CircularProgress$1,
    window: targetWindow
  });
  const bufferSubOverlayRightStyle = {};
  const bufferSubOverlayLeftStyle = {};
  const barSubOverlayRightStyle = {};
  const barSubOverlayLeftStyle = {};
  const buffer = (bufferValue - min) / (max - min) * 100;
  const progress = (value - min) / (max - min) * 100;
  if (progress <= 50) {
    const rotationAngle = getRotationAngle(progress);
    barSubOverlayRightStyle.transform = `rotate(${rotationAngle}deg)`;
    barSubOverlayLeftStyle.transform = "rotate(-180deg)";
  } else {
    const rotationAngle = getRotationAngle(progress, 50);
    barSubOverlayRightStyle.transform = "rotate(0deg)";
    barSubOverlayLeftStyle.transform = `rotate(${rotationAngle}deg)`;
  }
  if (buffer <= 50) {
    const rotationAngle = getRotationAngle(buffer);
    bufferSubOverlayRightStyle.transform = `rotate(${rotationAngle}deg)`;
    bufferSubOverlayLeftStyle.transform = "rotate(-180deg)";
  } else {
    const rotationAngle = getRotationAngle(buffer, 50);
    bufferSubOverlayRightStyle.transform = "rotate(0deg)";
    bufferSubOverlayLeftStyle.transform = `rotate(${rotationAngle}deg)`;
  }
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      className: clsx.clsx(withBaseName(), className),
      ref,
      role: "progressbar",
      "aria-valuemax": max,
      "aria-valuemin": min,
      "aria-valuenow": Math.round(value),
      ...rest,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("track") }),
        /* @__PURE__ */ jsxRuntime.jsxs("div", { className: withBaseName("bars"), children: [
          buffer > 0 && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferOverlayRight"), children: /* @__PURE__ */ jsxRuntime.jsxs(
            "div",
            {
              className: clsx.clsx(withBaseName("bufferSubOverlayRight"), {
                [withBaseName("bufferSubOverlay")]: buffer <= 50
              }),
              style: bufferSubOverlayRightStyle,
              children: [
                /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferBackground") }),
                /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferBorder") })
              ]
            }
          ) }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("barOverlayRight"), children: /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: withBaseName("barSubOverlayRight"),
              style: barSubOverlayRightStyle,
              children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bar") })
            }
          ) }),
          buffer > 0 && /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferOverlayLeft"), children: /* @__PURE__ */ jsxRuntime.jsxs(
            "div",
            {
              className: clsx.clsx(
                withBaseName("bufferSubOverlay"),
                withBaseName("bufferSubOverlayLeft")
              ),
              style: bufferSubOverlayLeftStyle,
              children: [
                /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferBorder") }),
                /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bufferBackground") })
              ]
            }
          ) }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("barOverlayLeft"), children: /* @__PURE__ */ jsxRuntime.jsx(
            "div",
            {
              className: withBaseName("barSubOverlayLeft"),
              style: barSubOverlayLeftStyle,
              children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: withBaseName("bar") })
            }
          ) })
        ] }),
        !hideLabel && /* @__PURE__ */ jsxRuntime.jsx(Text.Text, { styleAs: "h2", className: withBaseName("progressLabel"), children: `${Math.round(progress)} %` })
      ]
    }
  );
});

exports.CircularProgress = CircularProgress;
