// @salt-ds/core/dist-es/pill/Pill.js
import { jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { forwardRef } from 'react';
import '../button/Button.js';
import { useButton } from '../button/useButton.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Pill.css.js';

const withBaseName = makePrefixer("saltPill");
const Pill = forwardRef(function Pill2({
  children,
  className,
  disabled,
  onKeyUp,
  onKeyDown,
  onClick,
  onBlur,
  ...rest
}, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-pill",
    css: css_248z,
    window: targetWindow
  });
  const { buttonProps, active } = useButton({
    disabled,
    onKeyUp,
    onKeyDown,
    onClick,
    onBlur
  });
  const { tabIndex, ...restButtonProps } = buttonProps;
  return /* @__PURE__ */ jsx(
    "button",
    {
      "data-testid": "pill",
      ref,
      className: clsx(
        withBaseName(),
        withBaseName("clickable"),
        { [withBaseName("active")]: active },
        className
      ),
      type: "button",
      ...restButtonProps,
      ...rest,
      children
    }
  );
});

export { Pill };
