// @salt-ds/core/dist-es/badge/Badge.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import css_248z from './Badge.css.js';

const withBaseName = makePrefixer("saltBadge");
const Badge = forwardRef(function Badge2({ value, max = 999, className, children, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-badge",
    css: css_248z,
    window: targetWindow
  });
  const valueText = typeof value === "number" && value > max ? `${max}+` : value;
  const dotBadge = typeof value === "undefined";
  return /* @__PURE__ */ jsxs("span", { className: clsx(withBaseName(), className), ref, ...rest, children: [
    children,
    /* @__PURE__ */ jsx(
      "span",
      {
        className: clsx(withBaseName("badge"), {
          [withBaseName("topRight")]: children,
          [withBaseName("dotBadge")]: dotBadge
        }),
        children: valueText
      }
    )
  ] });
});

export { Badge };
