// @salt-ds/core/dist-es/slider/internal/SliderTrack.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { Text } from '../../text/Text.js';
import '../../text/Code.js';
import '../../text/Display.js';
import '../../text/Headings.js';
import '../../text/Label.js';
import '../../text/TextAction.js';
import '../../text/TextNotation.js';
import { makePrefixer } from '../../utils/makePrefixer.js';
import '../../utils/useFloatingUI/useFloatingUI.js';
import '../../utils/useId.js';
import '../../salt-provider/SaltProvider.js';
import '../../viewport/ViewportProvider.js';
import css_248z from './SliderTrack.css.js';
import { calculateMarkPosition, calculatePercentage } from './utils.js';

const withBaseName = makePrefixer("saltSliderTrack");
const SliderTrack = forwardRef(
  function SliderTrack2({
    children,
    className,
    constrainLabelPosition = false,
    disabled,
    showTicks,
    format,
    handlePointerDown,
    isDragging,
    isRange = false,
    marks,
    max,
    maxLabel,
    min,
    minLabel,
    progressPercentage = 0,
    progressPercentageRange = [0, 0],
    sliderRef,
    ...rest
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-slider-track",
      css: css_248z,
      window: targetWindow
    });
    const checkIsMarkSelected = (value) => {
      const markPercentage = calculatePercentage(value, max, min);
      if (isRange) {
        return markPercentage > progressPercentageRange[0] && markPercentage < progressPercentageRange[1];
      }
      return markPercentage < progressPercentage;
    };
    const checkIsMarkOverlapped = (value) => {
      const markPercentage = calculatePercentage(value, max, min);
      if (isRange) {
        return markPercentage === progressPercentageRange[0] || markPercentage === progressPercentageRange[1];
      }
      return markPercentage === progressPercentage;
    };
    const hasMinTick = () => {
      return (marks == null ? void 0 : marks.some((mark) => mark.value === min)) || false;
    };
    const hasMaxTick = () => {
      return (marks == null ? void 0 : marks.some((mark) => mark.value === max)) || false;
    };
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: clsx(withBaseName(), className, {
          [withBaseName("disabled")]: disabled,
          [withBaseName("dragging")]: isDragging,
          [withBaseName("range")]: isRange,
          [withBaseName("withMarks")]: marks,
          [withBaseName("constrainLabelPosition")]: constrainLabelPosition,
          [withBaseName("withTicks")]: showTicks
        }),
        "data-testid": "sliderTrack",
        ref,
        ...rest,
        children: /* @__PURE__ */ jsxs("div", { className: clsx(withBaseName("container")), children: [
          minLabel && /* @__PURE__ */ jsx(
            Text,
            {
              "aria-hidden": true,
              className: withBaseName("minLabel"),
              color: "secondary",
              disabled,
              styleAs: "label",
              children: minLabel || (format == null ? void 0 : format(min))
            }
          ),
          /* @__PURE__ */ jsxs(
            "div",
            {
              onPointerDown: handlePointerDown,
              className: withBaseName("wrapper"),
              children: [
                /* @__PURE__ */ jsxs(
                  "div",
                  {
                    className: clsx(withBaseName("rail"), {
                      [withBaseName("hasMinTick")]: hasMinTick() && showTicks,
                      [withBaseName("hasMaxTick")]: hasMaxTick() && showTicks
                    }),
                    ref: sliderRef,
                    style: {
                      ...progressPercentage !== void 0 && {
                        "--slider-progressPercentage": `${progressPercentage}%`
                      },
                      ...(progressPercentageRange == null ? void 0 : progressPercentageRange[0]) !== void 0 && {
                        "--slider-progressPercentageStart": `${progressPercentageRange[0]}%`
                      },
                      ...(progressPercentageRange == null ? void 0 : progressPercentageRange[1]) !== void 0 && {
                        "--slider-progressPercentageEnd": `${progressPercentageRange[1]}%`
                      }
                    },
                    children: [
                      isRange && /* @__PURE__ */ jsx("div", { className: clsx(withBaseName("fill")) }),
                      children
                    ]
                  }
                ),
                marks && showTicks && /* @__PURE__ */ jsx("div", { className: withBaseName("ticks"), children: marks.map(({ value }) => /* @__PURE__ */ jsx(
                  "span",
                  {
                    style: {
                      left: `${calculateMarkPosition(value, max, min)}%`
                    },
                    className: clsx(
                      withBaseName("tick"),
                      {
                        [withBaseName("tickSelected")]: checkIsMarkSelected(value)
                      },
                      {
                        [withBaseName("tickHidden")]: checkIsMarkOverlapped(value)
                      }
                    )
                  },
                  `${value}-tick`
                )) }),
                marks && /* @__PURE__ */ jsx("div", { className: withBaseName("marks"), children: marks.map(({ label, value }) => /* @__PURE__ */ jsx(
                  "span",
                  {
                    "data-testid": "mark",
                    className: withBaseName("markLabel"),
                    style: {
                      left: `${calculateMarkPosition(value, max, min)}%`
                    },
                    children: label
                  },
                  `${value}-mark`
                )) })
              ]
            }
          ),
          maxLabel && /* @__PURE__ */ jsx(
            Text,
            {
              "aria-hidden": true,
              className: withBaseName("maxLabel"),
              color: "secondary",
              disabled,
              styleAs: "label",
              children: maxLabel || (format == null ? void 0 : format(max))
            }
          )
        ] })
      }
    );
  }
);

export { SliderTrack };
