// @salt-ds/core/dist-es/slider/internal/SliderTooltip.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import clsx from 'clsx';
import { Text } from '../../text/Text.js';
import '../../text/Code.js';
import '../../text/Display.js';
import '../../text/Headings.js';
import '../../text/Label.js';
import '../../text/TextAction.js';
import '../../text/TextNotation.js';
import 'react';
import { makePrefixer } from '../../utils/makePrefixer.js';
import '../../utils/useFloatingUI/useFloatingUI.js';
import '../../utils/useId.js';
import '../../salt-provider/SaltProvider.js';
import '../../viewport/ViewportProvider.js';
import css_248z from './SliderTooltip.css.js';

const withBaseName = makePrefixer("saltSliderTooltip");
const SliderTooltip = ({ value, open }) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-slider-tooltip",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "aria-hidden": true,
      className: clsx(withBaseName(), {
        [withBaseName("visible")]: open
      }),
      "data-testid": "sliderTooltip",
      children: [
        /* @__PURE__ */ jsxs(
          "svg",
          {
            className: withBaseName("arrow"),
            "aria-hidden": "true",
            viewBox: "0 1 14 14",
            children: [
              /* @__PURE__ */ jsx(
                "path",
                {
                  d: "M0,0 H14 L7,7 Q7,7 7,7 Z",
                  stroke: " var(--salt-container-primary-background)"
                }
              ),
              /* @__PURE__ */ jsx(
                "path",
                {
                  d: "M0,0 H14 L7,7 Q7,7 7,7 Z",
                  stroke: "var(--salt-container-primary-borderColor)"
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx(Text, { className: withBaseName("text"), children: value })
      ]
    }
  );
};

export { SliderTooltip };
