// @salt-ds/core/dist-es/slider/internal/SliderThumb.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { clsx } from 'clsx';
import { useState, useCallback, useEffect } from 'react';
import { makePrefixer } from '../../utils/makePrefixer.js';
import '../../utils/useFloatingUI/useFloatingUI.js';
import { useId } from '../../utils/useId.js';
import '../../salt-provider/SaltProvider.js';
import '../../viewport/ViewportProvider.js';
import css_248z from './SliderThumb.css.js';
import { SliderTooltip } from './SliderTooltip.js';

const withBaseName = makePrefixer("saltSliderThumb");
const SliderThumb = ({
  "aria-label": ariaLabel,
  "aria-valuetext": ariaValueText,
  "aria-labelledby": ariaLabelledBy,
  accessibleMaxText,
  accessibleMinText,
  disabled,
  format,
  handleInputChange,
  handleKeydownOnThumb,
  handlePointerDown,
  index = 0,
  inputRef,
  isFocusVisible,
  max,
  maxLabel,
  min,
  minLabel,
  offsetPercentage,
  restrictToMarks,
  showTooltip,
  sliderValue,
  step,
  stepMultiplier,
  trackDragging,
  ...rest
}) => {
  {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-slider-thumb",
      css: css_248z,
      window: targetWindow
    });
    const [isTooltipVisible, setIsTooltipVisible] = useState(false);
    const id = useId();
    const accessibleTextId = `saltSlider-${id}-${index}`;
    const value = Array.isArray(sliderValue) ? sliderValue[index] : sliderValue;
    const formattedValue = format ? format(value) : value;
    const handleKeyDown = useCallback((event) => {
      if (event.key === "Escape") {
        setIsTooltipVisible(false);
      }
    }, []);
    useEffect(() => {
      if (showTooltip && isTooltipVisible) {
        targetWindow == null ? void 0 : targetWindow.addEventListener("keydown", handleKeyDown);
      }
      return () => targetWindow == null ? void 0 : targetWindow.removeEventListener("keydown", handleKeyDown);
    }, [handleKeyDown, isTooltipVisible, showTooltip, targetWindow]);
    const handlePointerEnter = () => setIsTooltipVisible(true);
    const handlePointerLeave = () => {
      setTimeout(() => {
        setIsTooltipVisible(false);
      }, 300);
    };
    return /* @__PURE__ */ jsxs(
      "div",
      {
        className: clsx(withBaseName(), {
          [withBaseName("focusVisible")]: isFocusVisible,
          [withBaseName("disabled")]: disabled,
          [withBaseName("dragging")]: trackDragging,
          [withBaseName("secondThumb")]: index === 1
        }),
        "data-testid": "sliderThumb",
        onPointerDown: handlePointerDown,
        style: { left: offsetPercentage },
        ...showTooltip && {
          onPointerEnter: handlePointerEnter,
          onPointerLeave: handlePointerLeave
        },
        children: [
          showTooltip && /* @__PURE__ */ jsx(
            SliderTooltip,
            {
              value: formattedValue,
              open: (isTooltipVisible || trackDragging || isFocusVisible) && !disabled
            }
          ),
          /* @__PURE__ */ jsx(
            "input",
            {
              disabled,
              type: "range",
              ref: inputRef,
              className: withBaseName("input"),
              value,
              onChange: handleInputChange,
              onKeyDown: handleKeydownOnThumb,
              "aria-labelledby": ariaLabelledBy,
              "aria-valuenow": value,
              "aria-valuetext": ariaValueText || (format == null ? void 0 : format(value).toString()),
              "aria-label": ariaLabel,
              "aria-describedby": accessibleTextId,
              min,
              max,
              step,
              ...rest
            }
          ),
          /* @__PURE__ */ jsxs(
            "span",
            {
              "aria-hidden": "true",
              id: accessibleTextId,
              className: withBaseName("accessibleText"),
              children: [
                Array.isArray(sliderValue) && `${index === 0 ? "leading" : "trailing"}, ${(format == null ? void 0 : format(sliderValue[0])) || sliderValue[0]} to ${(format == null ? void 0 : format(sliderValue[1])) || sliderValue[1]}, `,
                "range",
                " ",
                accessibleMinText ? `${accessibleMinText} ${min}, ` : `minimum ${(format == null ? void 0 : format(min)) || min}, `,
                accessibleMaxText ? `${accessibleMaxText} ${max} ` : `maximum ${(format == null ? void 0 : format(max)) || max}`,
                restrictToMarks ? ", custom increments" : step !== 1 && `, increments of ${step}`
              ]
            }
          )
        ]
      }
    );
  }
};

export { SliderThumb };
