// @salt-ds/core/dist-cjs/badge/Badge.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var clsx = require('clsx');
var React = require('react');
var makePrefixer = require('../utils/makePrefixer.js');
require('../utils/useFloatingUI/useFloatingUI.js');
require('../utils/useId.js');
require('../salt-provider/SaltProvider.js');
require('../viewport/ViewportProvider.js');
var Badge$1 = require('./Badge.css.js');

const withBaseName = makePrefixer.makePrefixer("saltBadge");
const Badge = React.forwardRef(function Badge2({ value, max = 999, className, children, ...rest }, ref) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-badge",
    css: Badge$1,
    window: targetWindow
  });
  const valueText = typeof value === "number" && value > max ? `${max}+` : value;
  const dotBadge = typeof value === "undefined";
  return /* @__PURE__ */ jsxRuntime.jsxs("span", { className: clsx.clsx(withBaseName(), className), ref, ...rest, children: [
    children,
    /* @__PURE__ */ jsxRuntime.jsx(
      "span",
      {
        className: clsx.clsx(withBaseName("badge"), {
          [withBaseName("topRight")]: children,
          [withBaseName("dotBadge")]: dotBadge
        }),
        children: valueText
      }
    )
  ] });
});

exports.Badge = Badge;
